{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "source:\n",
    "\n",
    "NCBI viral genomes resource.\n",
    "Brister JR, Ako-Adjei D, Bao Y, Blinkova O. Nucleic Acids Res. 2015 Jan;43(Database issue):D571-7. doi: 10.1093/nar/gku1207. Epub 2014 Nov 26."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import bs4 as bs\n",
    "import requests\n",
    "import numpy as np\n",
    "import progressbar\n",
    "import os\n",
    "import csv\n",
    "from Bio import Entrez\n",
    "import re\n",
    "import time\n",
    "import sys"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Acquisition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Create a BeautifulSoup object for the main page\n",
    "url = 'https://www.ncbi.nlm.nih.gov/genomes/GenomesGroup.cgi'\n",
    "result = requests.get(url)\n",
    "main_soup = bs.BeautifulSoup(result.content, 'html.parser')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Parse the main_soup for the table of viruses\n",
    "table = main_soup.find('table', {'class': 'tblTxt'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# This table contains some subtables, which are included within single cells, called 'minitable' below\n",
    "# The rows in the minitables are formatted differently, but contain the same info that I want, so this function \n",
    "# will convert them into the same format as the rest. A list comprehension didn't work because the 'minirows' have \n",
    "# variable lengths\n",
    "def edit_minirow(row): \n",
    "    n = len(row)\n",
    "    #assuming edited rows should have structure =[row[0],row[2],'None','None', row[1],row[3], row[4]]\n",
    "    if n == 1:\n",
    "        new_minirow = [row[0], 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None']\n",
    "    elif n == 2:\n",
    "        new_minirow = [row[0], 'None', 'None', 'None', row[1], 'None', 'None', 'None', 'None', 'None', 'None']\n",
    "    elif n == 3:\n",
    "        new_minirow = [row[0], row[2], 'None', 'None', row[1], 'None', 'None', 'None', 'None', 'None', 'None']\n",
    "    elif n == 4:\n",
    "        new_minirow = [row[0], row[2], 'None', 'None', row[1], row[3], 'None', 'None', 'None', 'None', 'None']\n",
    "    elif n == 5:\n",
    "        new_minirow = [row[0], row[2], 'None', 'None', row[1], row[3], row[4], 'None', 'None', 'None', 'None']\n",
    "    else:\n",
    "        print('error:', row)\n",
    "    return(new_minirow)\n",
    "\n",
    "# Initialize the list of classifications\n",
    "class_list = ['Deltavirus', 'Genomoviridae', 'Retro-transcribing viruses', 'Satellites', \n",
    "'Virus families not assigned to an order', 'Virus-associated RNAs', 'dsDNA viruses, no RNA stage', 'dsRNA viruses',\n",
    "'environmental samples', 'ssDNA viruses', 'ssRNA viruses', 'unassigned viruses', 'unclassified RNA viruses',\n",
    "'unclassified archaeal viruses', 'unclassified bacterial viruses', 'unclassified virophages', \n",
    "'unclassified viruses']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "# Create an empty list of lists\n",
    "viral_genome_array = []\n",
    "\n",
    "# Create a list of rows that will ensure that for every row in a minitable, that row isn't copied again when the \n",
    "#  main table finds those rows. The alternative would be to not include a separate minitable code, but since the \n",
    "#  minitables are within cells, those cells' text is added to the table\n",
    "minirow_master_list = []\n",
    "\n",
    "# This is in order to add the previous non-minitable row to each minitable row, as it contains relevant info\n",
    "head_row = []\n",
    "\n",
    "# This variable will hold the subheader of the previous lighter colored column, which will contain extra info for the \n",
    "# light columns beneath it\n",
    "sub_head = ''\n",
    "sub_head_list = []\n",
    "\n",
    "# Go through all of the rows in the table \n",
    "for row in table.find_all('tr'):\n",
    "    this_row = []\n",
    "    # Go through each cell in each row\n",
    "    for cell in row.find_all('td'):\n",
    "        \n",
    "        # If the cell contains a table (the pulldown rows on the main page) reformat those rows like the normal rows\n",
    "        try:\n",
    "            minitable = cell.find('table')\n",
    "            for minirow in minitable.find_all('tr'):\n",
    "                this_minirow = []\n",
    "                for minicell in minirow.find_all('td'):\n",
    "                    if 'NC_' in minicell.text:\n",
    "                        this_minirow.append(minicell.find('a').attrs['href'])\n",
    "                    else:\n",
    "                        this_minirow.append(minicell.text.replace('\\n','').replace('\\xa0',''))\n",
    "                minirow_master_list.append(this_minirow)\n",
    "                edited_minirow = edit_minirow(this_minirow)\n",
    "                edited_minirow.append(head_row)\n",
    "                edited_minirow[0] = edited_minirow[0] + ' segment '\n",
    "                viral_genome_array.append(edited_minirow)\n",
    "                \n",
    "        # If there isn't a table, grab the href link for cells with a 'NC_', otherwise grab the text\n",
    "        except AttributeError:\n",
    "            if 'NC_' in cell.text:\n",
    "                this_row.append(cell.find('a').attrs['href'])\n",
    "            else:\n",
    "                this_row.append(cell.text.replace('\\n','').replace('\\xa0',''))\n",
    "    \n",
    "    # For the lighter rows, which share a subheader, either grab the subheader or append the current subheader in \n",
    "    # the 12th position (index 11)\n",
    "    try:\n",
    "        if row.attrs['bgcolor'] == '#F8F8F8':\n",
    "            if len(this_row) == 1:\n",
    "                # Classifications also have this lighter color, but I will to treat them differently to add them to \n",
    "                #  every row\n",
    "                if this_row[0] not in class_list:\n",
    "                    sub_head = this_row[0]\n",
    "                    sub_head_list.append(this_row)\n",
    "            else:\n",
    "                while len(this_row) < 11:\n",
    "                    this_row.append('None')\n",
    "                this_row.append(sub_head)\n",
    "    except:\n",
    "        pass\n",
    "    #\n",
    "    \n",
    "    # don't take rows that have already been taken down and reformatteed, and rows that include the text \n",
    "    # of a whole minitable (which have the name repeated)\n",
    "    if this_row not in minirow_master_list and this_row[0] not in this_row[1:]: \n",
    "        head_row = this_row \n",
    "        viral_genome_array.append(this_row)\n",
    "    \n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Add a classification and extra info feature in the first row (these will become dataframe column headers)\n",
    "viral_genome_array[0].append('Classification')\n",
    "viral_genome_array[0].append('Extra Info')\n",
    "\n",
    "# If a row contains only a classification, everything below that row falls into that category until the next \n",
    "# classification comes up\n",
    "classification = 'None'\n",
    "for row in viral_genome_array[1:]:\n",
    "    if row[0] in class_list:\n",
    "        classification = row[0]\n",
    "        viral_genome_array.remove(row)\n",
    "    else:\n",
    "        while len(row) < 11:\n",
    "            row.append('None')\n",
    "        row[10] = classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Gather a list of the head_rows that ended up in the extra info columns\n",
    "head_row_list = []\n",
    "for row in viral_genome_array[1:]:\n",
    "    if len(row) == 12:\n",
    "        head = row[11]\n",
    "        if head != 'None' and head not in head_row_list:\n",
    "            head_row_list.append(head)\n",
    "\n",
    "# now that I have a list of heads, go through the array again and remove rows that match those head_rows, as well\n",
    "#  as subheaders\n",
    "for row in viral_genome_array[1:]:\n",
    "    if row in head_row_list or row in sub_head_list:\n",
    "        viral_genome_array.remove(row)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Extract the info in 'Extra Info' lists \n",
    "# For each row in the array, if the row ends in a list, replace missing values in that row with the corresponding \n",
    "#  value in the list\n",
    "for row in viral_genome_array:\n",
    "    if type(row[-1]) == list:\n",
    "        extra_list = row[-1]\n",
    "        for i in range(len(row)):\n",
    "            if i not in [3,4,5,6]:\n",
    "                if row[i] == 'None' or row[i] == '' or row[i] == '-':\n",
    "                    row[i] = extra_list[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Number all of the segments\n",
    "segment_counts = {}\n",
    "\n",
    "for row in viral_genome_array:\n",
    "    name = row[0]\n",
    "    if 'segment' in name:\n",
    "        if name not in segment_counts:\n",
    "            segment_counts[name] = 1\n",
    "            row[0] += '1'\n",
    "        else:\n",
    "            segment_counts[name] += 1\n",
    "            row[0] += str(segment_counts[name])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Source information</th>\n",
       "      <th>Segm</th>\n",
       "      <th>Length</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Neighbors</th>\n",
       "      <th>Host</th>\n",
       "      <th>Created</th>\n",
       "      <th>Updated</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Extra Info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hepatitis delta virus</td>\n",
       "      <td>/nuccore/13277517</td>\n",
       "      <td></td>\n",
       "      <td>-</td>\n",
       "      <td>1682nt</td>\n",
       "      <td>2</td>\n",
       "      <td>320</td>\n",
       "      <td>vertebrates, human</td>\n",
       "      <td>04/29/1993</td>\n",
       "      <td>02/10/2015</td>\n",
       "      <td>Deltavirus</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Badger associated gemykibivirus 1</td>\n",
       "      <td>/nuccore/807743872</td>\n",
       "      <td>strain:588t</td>\n",
       "      <td>-</td>\n",
       "      <td>2112nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>04/16/2015</td>\n",
       "      <td>11/24/2015</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bemisia-associated genomovirus AdDF</td>\n",
       "      <td>/nuccore/1211677462</td>\n",
       "      <td>isolate:AdDF</td>\n",
       "      <td>-</td>\n",
       "      <td>2199nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/19/2017</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bemisia-associated genomovirus AdO</td>\n",
       "      <td>/nuccore/1211677465</td>\n",
       "      <td>isolate:AdO</td>\n",
       "      <td>-</td>\n",
       "      <td>2211nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/19/2017</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bemisia-associated genomovirus NfO</td>\n",
       "      <td>/nuccore/1211677468</td>\n",
       "      <td>isolate:NfO</td>\n",
       "      <td>-</td>\n",
       "      <td>2231nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/19/2017</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Genome            Accession  \\\n",
       "0                Hepatitis delta virus    /nuccore/13277517   \n",
       "1    Badger associated gemykibivirus 1   /nuccore/807743872   \n",
       "2  Bemisia-associated genomovirus AdDF  /nuccore/1211677462   \n",
       "3   Bemisia-associated genomovirus AdO  /nuccore/1211677465   \n",
       "4   Bemisia-associated genomovirus NfO  /nuccore/1211677468   \n",
       "\n",
       "  Source information Segm  Length Protein Neighbors                Host  \\\n",
       "0                       -  1682nt       2       320  vertebrates, human   \n",
       "1        strain:588t    -  2112nt       2         -         vertebrates   \n",
       "2       isolate:AdDF    -  2199nt       2         -       invertebrates   \n",
       "3        isolate:AdO    -  2211nt       2         -       invertebrates   \n",
       "4        isolate:NfO    -  2231nt       2         -       invertebrates   \n",
       "\n",
       "      Created     Updated Classification Extra Info  \n",
       "0  04/29/1993  02/10/2015     Deltavirus       None  \n",
       "1  04/16/2015  11/24/2015  Genomoviridae       None  \n",
       "2  06/27/2017  07/19/2017  Genomoviridae       None  \n",
       "3  06/27/2017  07/19/2017  Genomoviridae       None  \n",
       "4  06/27/2017  07/19/2017  Genomoviridae       None  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "viral_genomes = pd.DataFrame(viral_genome_array[2:], columns = viral_genome_array[0])#, columns = viral_genome_array[0])\n",
    "viral_genomes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------------------------------------------------\n",
    "##### This cell gets the full html 'soup' from each virus' page, which includes the genome. My intention is to only run this once since it will take a while, and save the result into a csv file in this directory. \n",
    "    length = len(viral_genomes)\n",
    "    \n",
    "    soup_dict = {}\n",
    "\n",
    "###### Using a progress bar, go through the dataframe, getting a soup for each genome and saving the result to a dict\n",
    "    with progressbar.ProgressBar(max_value=length) as bar:\n",
    "        for i in range(length):\n",
    "    \n",
    "            name = viral_genomes['Genome'][i]\n",
    "        \n",
    "            slug = viral_genomes['Accession'][i]\n",
    "        \n",
    "            result = requests.get('https://www.ncbi.nlm.nih.gov{}'.format(slug))\n",
    "        \n",
    "            soup = bs.BeautifulSoup(result.content, 'html.parser')\n",
    "        \n",
    "            soup_dict[name] = soup\n",
    "        \n",
    "            bar.update(i)\n",
    "        \n",
    "    os.system('say \"download complete\"')\n",
    "\n",
    "###### Write the dict to a csv: \"virus_soups.csv\"\n",
    "    w = csv.writer(open(\"virus_soups.csv\", \"w\"))\n",
    "    for key, val in soup_dict.items():\n",
    "        w.writerow([key, val])\n",
    "----------------------------------------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>BeautifulSoup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hepatitis delta virus</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Badger associated gemykibivirus 1</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bemisia-associated genomovirus AdDF</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bemisia-associated genomovirus AdO</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bemisia-associated genomovirus NfO</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Name  \\\n",
       "0                Hepatitis delta virus   \n",
       "1    Badger associated gemykibivirus 1   \n",
       "2  Bemisia-associated genomovirus AdDF   \n",
       "3   Bemisia-associated genomovirus AdO   \n",
       "4   Bemisia-associated genomovirus NfO   \n",
       "\n",
       "                                       BeautifulSoup  \n",
       "0  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...  \n",
       "1  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...  \n",
       "2  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...  \n",
       "3  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...  \n",
       "4  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load in the dataframe including the BeautifulSoups\n",
    "meta = pd.read_csv('virus_soups.csv', header = None, names = ['Name', 'BeautifulSoup'])\n",
    "meta['BeautifulSoup'] = meta['BeautifulSoup'].apply(lambda x: bs.BeautifulSoup(x, 'html.parser'))\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# This function returns the id that I need for each virus to access it in the NCBI Entrez database \n",
    "def get_db_id (soup):\n",
    "    try:\n",
    "        line = soup.find('p', {'class': 'itemid'}).text\n",
    "        itemid = (line[25:])\n",
    "        return(itemid)\n",
    "    except AttributeError:\n",
    "        return(np.nan)\n",
    "\n",
    "# Add the sequence type to the main DataFrame\n",
    "def get_seq_type(soup):\n",
    "    try:\n",
    "        head = soup.find('div', {\"class\":\"rprtheader\"})\n",
    "        line = head.find('h1')\n",
    "        title = (line.text)\n",
    "        sequence_type = re.search(',[a-z,0-9, ]+', title)[0]\n",
    "        return(sequence_type[2:])\n",
    "    except AttributeError:\n",
    "        return('error')\n",
    "    except TypeError:\n",
    "        return('error')    \n",
    "    \n",
    "meta['ItemId'] = meta['BeautifulSoup'].apply(get_db_id)\n",
    "viral_genomes['SequenceType'] = meta['BeautifulSoup'].apply(get_seq_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 9461 entries, 0 to 9460\n",
      "Data columns (total 3 columns):\n",
      "Name             9461 non-null object\n",
      "BeautifulSoup    9461 non-null object\n",
      "ItemId           9459 non-null object\n",
      "dtypes: object(3)\n",
      "memory usage: 221.8+ KB\n"
     ]
    }
   ],
   "source": [
    "# Check to make sure we have an itemid for each virus/segment\n",
    "meta.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4240\n",
      "6110\n"
     ]
    }
   ],
   "source": [
    "# identify missing ids (np.nan is a float type)\n",
    "for i in range(len(meta['ItemId'])):\n",
    "    string = meta['ItemId'][i]\n",
    "    if type(string) == float:\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#manually enter ids for 4240, 6110\n",
    "\n",
    "meta['ItemId'][4240] = 'NC_032149.2'\n",
    "meta['ItemId'][6110] = 'NC_025484.1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "complete genome                                                                                                                                  5532\n",
       "complete sequence                                                                                                                                1683\n",
       "complete cds                                                                                                                                     1409\n",
       "error                                                                                                                                             201\n",
       "hypothetical protein 2, and hypothetical protein 3 genes, complete cds                                                                             93\n",
       "                                                                                                                                                   63\n",
       "segment                                                                                                                                            54\n",
       "isolate                                                                                                                                            35\n",
       "complete genome, isolate                                                                                                                           33\n",
       "partial genome                                                                                                                                     31\n",
       "hypothetical protein 2, hypothetical protein 3, and hypothetical protein 4 genes, complete cds                                                     27\n",
       "complete genome, strain                                                                                                                            16\n",
       "genomic                                                                                                                                            11\n",
       "genome                                                                                                                                             11\n",
       "strain                                                                                                                                              9\n",
       "complete genome, isolate                                                                                                                            8\n",
       "complete cds, strain                                                                                                                                8\n",
       "hypothetical protein, putative structural protein, and hypothetical protein genes, complete cds                                                     7\n",
       "partial cds                                                                                                                                         7\n",
       "putative phosphoprotein                                                                                                                             7\n",
       "hypothetical protein 2, hypothetical protein 3, putative glycoprotein, and                                                                          6\n",
       "gene, complete cds                                                                                                                                  6\n",
       "putative glycoprotein, and hypothetical protein genes, complete cds                                                                                 6\n",
       "phosphoprotein, hypothetical proteins, matrix, glycoprotein, hypothetical protein, and polymerase genes, complete cds                               6\n",
       "complete cds, isolate                                                                                                                               5\n",
       "phosphoprotein                                                                                                                                      5\n",
       "phosphoprotein, matrix, glycoprotein, and polymerase genes, complete cds                                                                            5\n",
       "hypothetical protein, and                                                                                                                           5\n",
       "hypothetical protein 2, hypothetical protein 3, hypothetical protein 4, and hypothetical protein 5 genes, complete cds                              4\n",
       "hypothetical protein 1, and hypothetical protein 2 genes, complete cds                                                                              4\n",
       "                                                                                                                                                 ... \n",
       "putative readthrough protein, and coat protein genes, complete cds                                                                                  1\n",
       "hypothetical protein                                                                                                                                1\n",
       "capsid protein and movement protein                                                                                                                 1\n",
       "segment 1, complete sequence, isolate                                                                                                               1\n",
       "complete segment 1, genomic                                                                                                                         1\n",
       "contig00001 sequence                                                                                                                                1\n",
       "complete segment 2, genomic                                                                                                                         1\n",
       "phosphoprotein, hypothetical protein, matrix, hypothetical protein, glycoprotein, hypothetical protein, and polymerase genes, complete cds          1\n",
       "hypothetical protein 2, putative                                                                                                                    1\n",
       "phosphoprotein, hypothetical proteins, matrix, hypothetical proteins, glycoprotein, hypothetical proteins, and polymerase genes, complete cds       1\n",
       "triple gene block protein 1                                                                                                                         1\n",
       "hypothetical protein, putative                                                                                                                      1\n",
       "putative major structural protein, and putative structural protein genes, complete cds                                                              1\n",
       "nearly complete genome, strain                                                                                                                      1\n",
       "putative                                                                                                                                            1\n",
       "nonstructural protein 1                                                                                                                             1\n",
       "triple gene block                                                                                                                                   1\n",
       "membrane                                                                                                                                            1\n",
       "hypothetical protein 2, and putative capsid protein genes, complete cds                                                                             1\n",
       "segment 3, complete sequence                                                                                                                        1\n",
       "segment 2, complete sequence, isolate                                                                                                               1\n",
       "hypothetical protein, putative matrix protein, putative glycoprotein, and                                                                           1\n",
       "segment 2, complete sequence, strain                                                                                                                1\n",
       "complate genome                                                                                                                                     1\n",
       "hypothetical protein, putative glycoprotein, and hypothetical protein genes, complete cds                                                           1\n",
       "isolate 1                                                                                                                                           1\n",
       "putative replication protein, and putative capsid protein genes, complete cds                                                                       1\n",
       "genes, complete cds                                                                                                                                 1\n",
       "hypothetical protein, and putative capsid protein genes, complete cds                                                                               1\n",
       "hypothetical protein, putative matrix protein, putative glycoprotein 1,                                                                             1\n",
       "Name: SequenceType, Length: 151, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "viral_genomes['SequenceType'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### As with scraping the html for each virus above, I will attempt to only extract the report for each virus once, \n",
    "###### I will save this to the meta df and thereafter I'll be importing that and not running this cell. \n",
    "\n",
    "    meta['Report'] = 'None'\n",
    "    length = len(meta['ItemId'])\n",
    "\n",
    "    with progressbar.ProgressBar(max_value=length) as bar:\n",
    "        for i in range(length):     #starting at 4235 because it stopped at 4240\n",
    "            try:\n",
    "                Entrez.email = 'qdupupet@umass.edu'\n",
    "                handle = Entrez.efetch(db=\"nuccore\", id=meta['ItemId'][i], rettype=\"gb\", retmode=\"text\")\n",
    "                report = handle.read()\n",
    "                meta['Report'][i] = report\n",
    "            except AttributeErrror:\n",
    "                pass\n",
    "            time.sleep(1)\n",
    "            bar.update(i)\n",
    "        \n",
    "    meta.to_csv('virus_meta.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### I didn't get all the reports, so I'll try getting the rest with a slightly different method\n",
    "\n",
    "##### just until I have the final version\n",
    "##### Read in meta csv with reports 0-4240 already in\n",
    "    chunksize = 1000\n",
    "    TextFileReader = pd.read_csv('virus_meta.csv', chunksize=chunksize, iterator=True)\n",
    "    meta = pd.concat(TextFileReader, ignore_index=True)\n",
    "\n",
    "    meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### The extra reports will be loaded into a separate csv (writing to meta took too long) and the resuslts will be added to 'meta'\n",
    "    reports = []\n",
    "    length = len(meta['ItemId'])\n",
    "\n",
    "    with progressbar.ProgressBar(max_value=length) as bar:\n",
    "        for i in range(4235, length):\n",
    "            line = [i]\n",
    "            try:\n",
    "                Entrez.email = 'qdupupet@umass.edu'\n",
    "                handle = Entrez.efetch(db=\"nuccore\", id=meta['ItemId'][i], rettype=\"gb\", retmode=\"text\")\n",
    "                report = handle.read()\n",
    "                line.append(report)\n",
    "            except AttributeErrror:\n",
    "                pass\n",
    "            reports.append(line)\n",
    "            time.sleep(1)\n",
    "            bar.update(i)\n",
    "\n",
    "    reports = pd.DataFrame(reports)\n",
    "    reports.to_csv('virus_extra_reports.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Add the batch of reports to the subset of the meta csv that they blong to \n",
    "    extra_report_dict = {}\n",
    "    with open('virus_extra_reports.csv', 'r') as csvfile:\n",
    "        report_reader = csv.reader(csvfile, delimiter=',', quotechar='\"')\n",
    "        for row in report_reader:\n",
    "            extra_report_dict[row[1]] = row[2]\n",
    "\n",
    "    submeta = meta[:][4236:7093]\n",
    "\n",
    "    submeta['Report'] = submeta['Unnamed: 0'].apply(lambda x: extra_report_dict[str(x)])\n",
    "\n",
    "    submeta.tail()\n",
    "\n",
    "    meta.to_csv('virus_meta.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Name</th>\n",
       "      <th>BeautifulSoup</th>\n",
       "      <th>ItemId</th>\n",
       "      <th>Report</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Hepatitis delta virus</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\\n&lt;!DOCT...</td>\n",
       "      <td>NC_001653.2</td>\n",
       "      <td>LOCUS       NC_001653               1682 bp ss...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Badger associated gemykibivirus 1</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\\n&lt;!DOCT...</td>\n",
       "      <td>NC_026806.1</td>\n",
       "      <td>LOCUS       NC_026806               2112 bp   ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Bemisia-associated genomovirus AdDF</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\\n&lt;!DOCT...</td>\n",
       "      <td>NC_035137.1</td>\n",
       "      <td>LOCUS       NC_035137               2199 bp   ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Bemisia-associated genomovirus AdO</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\\n&lt;!DOCT...</td>\n",
       "      <td>NC_035138.1</td>\n",
       "      <td>LOCUS       NC_035138               2211 bp   ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Bemisia-associated genomovirus NfO</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\\n&lt;!DOCT...</td>\n",
       "      <td>NC_035139.1</td>\n",
       "      <td>LOCUS       NC_035139               2231 bp   ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0                                 Name  \\\n",
       "0           0                Hepatitis delta virus   \n",
       "1           1    Badger associated gemykibivirus 1   \n",
       "2           2  Bemisia-associated genomovirus AdDF   \n",
       "3           3   Bemisia-associated genomovirus AdO   \n",
       "4           4   Bemisia-associated genomovirus NfO   \n",
       "\n",
       "                                       BeautifulSoup       ItemId  \\\n",
       "0  <?xml version=\"1.0\" encoding=\"utf-8\"?>\\n<!DOCT...  NC_001653.2   \n",
       "1  <?xml version=\"1.0\" encoding=\"utf-8\"?>\\n<!DOCT...  NC_026806.1   \n",
       "2  <?xml version=\"1.0\" encoding=\"utf-8\"?>\\n<!DOCT...  NC_035137.1   \n",
       "3  <?xml version=\"1.0\" encoding=\"utf-8\"?>\\n<!DOCT...  NC_035138.1   \n",
       "4  <?xml version=\"1.0\" encoding=\"utf-8\"?>\\n<!DOCT...  NC_035139.1   \n",
       "\n",
       "                                              Report  \n",
       "0  LOCUS       NC_001653               1682 bp ss...  \n",
       "1  LOCUS       NC_026806               2112 bp   ...  \n",
       "2  LOCUS       NC_035137               2199 bp   ...  \n",
       "3  LOCUS       NC_035138               2211 bp   ...  \n",
       "4  LOCUS       NC_035139               2231 bp   ...  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in the meta with reports 0-7092 already in \n",
    "chunksize = 1000\n",
    "TextFileReader = pd.read_csv('virus_meta.csv', chunksize=chunksize, iterator=True)\n",
    "meta = pd.concat(TextFileReader, ignore_index=True)\n",
    "\n",
    "meta[7090:7095]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### The extra reports will be loaded into a separate csv (writing to meta took too long) and the resuslts will be added to 'meta'\n",
    "    reports2 = []\n",
    "    length = len(meta['ItemId'])\n",
    "\n",
    "    with progressbar.ProgressBar(max_value=length) as bar:\n",
    "        for i in range(7090, length):\n",
    "            line = [i]\n",
    "            try:\n",
    "                Entrez.email = 'qdupupet@umass.edu'\n",
    "                handle = Entrez.efetch(db=\"nuccore\", id=meta['ItemId'][i], rettype=\"gb\", retmode=\"text\")\n",
    "                report = handle.read()\n",
    "                line.append(report)\n",
    "            except AttributeErrror:\n",
    "                pass\n",
    "            reports2.append(line)\n",
    "            time.sleep(1)\n",
    "            bar.update(i)\n",
    "\n",
    "    reports2 = pd.DataFrame(reports2)\n",
    "    reports2.to_csv('virus_extra_reports2.csv', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Write the last batch of reports to a dictionary and enter them into the DataFrame subset starting at index 7090\n",
    "    csv.field_size_limit(sys.maxsize)\n",
    "    \n",
    "    extra_report_dict_2 = {}\n",
    "    with open('virus_extra_reports2.csv', 'r') as csvfile:\n",
    "        report2_reader = csv.reader(csvfile, delimiter=',', quotechar='\"')\n",
    "        for row in report2_reader:\n",
    "            extra_report_dict_2[row[0]] = row[1]\n",
    "\n",
    "    submeta2 = meta[:][7090:]\n",
    "\n",
    "    submeta2['Report'] = submeta2['Unnamed: 0'].apply(lambda x: extra_report_dict_2[str(x)])\n",
    "\n",
    "    submeta2.tail()\n",
    "\n",
    "    meta.to_csv('full_virus_meta.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Name</th>\n",
       "      <th>BeautifulSoup</th>\n",
       "      <th>ItemId</th>\n",
       "      <th>Report</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9456</th>\n",
       "      <td>9456</td>\n",
       "      <td>Wilkie partiti-like virus 1</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\\n&lt;!DOCT...</td>\n",
       "      <td>NC_035122.1</td>\n",
       "      <td>LOCUS       NC_035122               2100 bp   ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9457</th>\n",
       "      <td>9457</td>\n",
       "      <td>Wilkie partiti-like virus 2</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\\n&lt;!DOCT...</td>\n",
       "      <td>NC_035119.1</td>\n",
       "      <td>LOCUS       NC_035119               1819 bp   ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9458</th>\n",
       "      <td>9458</td>\n",
       "      <td>Wolkberg virus segment 1</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\\n&lt;!DOCT...</td>\n",
       "      <td>NC_034633.1</td>\n",
       "      <td>LOCUS       NC_034633               6873 bp   ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9459</th>\n",
       "      <td>9459</td>\n",
       "      <td>Wolkberg virus segment 2</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\\n&lt;!DOCT...</td>\n",
       "      <td>NC_034631.1</td>\n",
       "      <td>LOCUS       NC_034631               4461 bp   ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9460</th>\n",
       "      <td>9460</td>\n",
       "      <td>Wolkberg virus segment 3</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\\n&lt;!DOCT...</td>\n",
       "      <td>NC_034632.1</td>\n",
       "      <td>LOCUS       NC_034632                978 bp   ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0                         Name  \\\n",
       "9456        9456  Wilkie partiti-like virus 1   \n",
       "9457        9457  Wilkie partiti-like virus 2   \n",
       "9458        9458     Wolkberg virus segment 1   \n",
       "9459        9459     Wolkberg virus segment 2   \n",
       "9460        9460     Wolkberg virus segment 3   \n",
       "\n",
       "                                          BeautifulSoup       ItemId  \\\n",
       "9456  <?xml version=\"1.0\" encoding=\"utf-8\"?>\\n<!DOCT...  NC_035122.1   \n",
       "9457  <?xml version=\"1.0\" encoding=\"utf-8\"?>\\n<!DOCT...  NC_035119.1   \n",
       "9458  <?xml version=\"1.0\" encoding=\"utf-8\"?>\\n<!DOCT...  NC_034633.1   \n",
       "9459  <?xml version=\"1.0\" encoding=\"utf-8\"?>\\n<!DOCT...  NC_034631.1   \n",
       "9460  <?xml version=\"1.0\" encoding=\"utf-8\"?>\\n<!DOCT...  NC_034632.1   \n",
       "\n",
       "                                                 Report  \n",
       "9456  LOCUS       NC_035122               2100 bp   ...  \n",
       "9457  LOCUS       NC_035119               1819 bp   ...  \n",
       "9458  LOCUS       NC_034633               6873 bp   ...  \n",
       "9459  LOCUS       NC_034631               4461 bp   ...  \n",
       "9460  LOCUS       NC_034632                978 bp   ...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in the full meta df with all reports.\n",
    "chunksize = 1000\n",
    "TextFileReader = pd.read_csv('full_virus_meta.csv', chunksize=chunksize, iterator=True)\n",
    "meta = pd.concat(TextFileReader, ignore_index=True)\n",
    "\n",
    "meta.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error at LOCUS       NC_016763              43221 bp    DNA     circular PHG 08-DEC-2014\n",
      "DEFINITION  Salmonella phage SE2, complete genome.\n",
      "ACCESSION   NC_016763\n",
      "VERSION     NC_016763.1\n",
      "DBLINK      BioProject: PRJNA82643\n",
      "KEYWORDS    RefSeq.\n",
      "SOURCE      Salmonella phage SE2\n",
      "  ORGANISM  Salmonella phage SE2\n",
      "            Viruses; dsDNA viruses, no RNA stage; Caudovirales; Siphoviridae;\n",
      "            Guernseyvirinae; Jerseyvirus.\n",
      "REFERENCE   1  (bases 1 to 43221)\n",
      "  AUTHORS   Tiwari,B.R., Kim,S. and Kim,J.\n",
      "  TITLE     Complete Genomic Sequence of Salmonella enterica Serovar\n",
      "            Enteritidis Phage SE2\n",
      "  JOURNAL   J. Virol. 86 (14), 7712 (2012)\n",
      "   PUBMED   22733878\n",
      "REFERENCE   2  (bases 1 to 43221)\n",
      "  CONSRTM   NCBI Genome Project\n",
      "  TITLE     Direct Submission\n",
      "  JOURNAL   Submitted (10-FEB-2012) National Center for Biotechnology\n",
      "            Information, NIH, Bethesda, MD 20894, USA\n",
      "REFERENCE   3  (bases 1 to 43221)\n",
      "  AUTHORS   Tiwari,B.R., Kim,S. and Kim,J.\n",
      "  TITLE     Direct Submission\n",
      "  JOURNAL   Submitted (09-NOV-2011) Department of Microbiology, Kyungpook\n",
      "            National University, School of Medicine, 101 Dongin-dong, Jung-gu,\n",
      "            Daegu, Kyungpook 700-422, South Korea\n",
      "COMMENT     PROVISIONAL REFSEQ: This record has not yet been subject to final\n",
      "            NCBI review. The reference sequence is identical to JQ007353.\n",
      "            COMPLETENESS: full length.\n",
      "FEATURES             Location/Qualifiers\n",
      "     source          1..43221\n",
      "                     /organism=\"Salmonella phage SE2\"\n",
      "                     /mol_type=\"genomic DNA\"\n",
      "                     /isolation_source=\"sewage sample\"\n",
      "                     /host=\"Salmonella enterica serovar Enteritidis\"\n",
      "                     /db_xref=\"taxon:1115478\"\n",
      "                     /country=\"South Korea\"\n",
      "                     /collection_date=\"10-Mar-2011\"\n",
      "     gene            complement(100..1524)\n",
      "                     /locus_tag=\"SaPhSE2_gp01\"\n",
      "                     /db_xref=\"GeneID:11669859\"\n",
      "     CDS             complement(100..1524)\n",
      "                     /locus_tag=\"SaPhSE2_gp01\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=11\n",
      "                     /product=\"hypothetical protein\"\n",
      "                     /protein_id=\"YP_005098097.1\"\n",
      "                     /db_xref=\"GeneID:11669859\"\n",
      "                     /translation=\"MSKFRRREYQKIMTSFMLQHPRCNIWCGMGGGKTSSTMWVLSRL\n",
      "                     FRNGQLTEEDRVLILAPLRVASGTWPAEQTKWGFPCLRVIDATGSEKRRIAALESDAN\n",
      "                     VVCTNYEVIEWLIDYYGKDDWPFTVIVADESTKLKSFRSRSGGSKRAKALSKVAFGKV\n",
      "                     KRFINLTGTPSPNGLKDLWGQNWFIDAGERLGSSYTAFTDRWFNSVQKGKSAMAREYH\n",
      "                     ARPGADNEIHQKMKDISLTIDAAEWFGCEAPVIVPVEIDLPKKARQAYIDMEEELFAE\n",
      "                     LESGEVEAANAAAKTAKCLQIASGAVYVSGPDGEATKDWEKVHDAKLDALESIVEELQ\n",
      "                     GAPLLVAYQFKHELERILRRFPQAQAFAKGAKGNKQMESWNRGEIEILCVHPASAGHG\n",
      "                     LNLQDGGHHLAFISQGWNLEHYLQVVERIGPVRQKQAGHERPVFLYHIVAKDTLDEVV\n",
      "                     AARTDEKKSVQEELLNYMKRRGKR\"\n",
      "     gene            complement(1521..2018)\n",
      "                     /locus_tag=\"SaPhSE2_gp02\"\n",
      "                     /db_xref=\"GeneID:11669860\"\n",
      "     CDS             complement(1521..2018)\n",
      "                     /locus_tag=\"SaPhSE2_gp02\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=11\n",
      "                     /product=\"hypothetical protein\"\n",
      "                     /protein_id=\"YP_005098098.1\"\n",
      "                     /db_xref=\"GeneID:11669860\"\n",
      "                     /translation=\"MSKFNRRPDIEKAKRLLNYDPETGIFTRKVSIKGRNAGEVAGGP\n",
      "                     NDKGYIVITVSGVRIKAHHLAWAFVYGEYHNGELDHKDRNRANNAINNIRPATRSQQI\n",
      "                     QNRDCSSHNTSGAIGVYQIPSGRWRARIGVNNKYIHLGYFDTIEEASRVYQRAAEIYF\n",
      "                     GEFKA\"\n",
      "     gene            complement(2015..2206)\n",
      "                     /locus_tag=\"SaPhSE2_gp03\"\n",
      "                     /db_xref=\"GeneID:11669861\"\n",
      "     CDS             complement(2015..2206)\n",
      "                     /locus_tag=\"SaPhSE2_gp03\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=11\n",
      "                     /product=\"hypothetical protein\"\n",
      "                     /protein_id=\"YP_005098099.1\"\n",
      "                     /db_xref=\"GeneID:11669861\"\n",
      "                     /translation=\"MKYEYDRKPARDIVPGDMIFNVKTRQPVAVDTVFVESNGKLVIE\n",
      "                     DVTGNVTAFGRKELVLVLK\"\n",
      "     gene            complement(2238..2432)\n",
      "                     /locus_tag=\"SaPhSE2_gp04\"\n",
      "                     /db_xref=\"GeneID:11669862\"\n",
      "     CDS             complement(2238..2432)\n",
      "                     /locus_tag=\"SaPhSE2_gp04\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=11\n",
      "                     /product=\"hypothetical protein\"\n",
      "                     /protein_id=\"YP_005098100.1\"\n",
      "                     /db_xref=\"GeneID:11669862\"\n",
      "                     /translation=\"MGAPDLLVILPGGIIWFVEVKKDENTKPDPHQLREHERMRKRGA\n",
      "                     NVFVVGSKKQVDKLIEHYYI\"\n",
      "     gene            complement(2613..3215)\n",
      "                     /locus_tag=\"SaPhSE2_gp05\"\n",
      "                     /db_xref=\"GeneID:11669863\"\n",
      "     CDS             complement(2613..3215)\n",
      "                     /locus_tag=\"SaPhSE2_gp05\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=11\n",
      "                     /product=\"DNA polymerase\"\n",
      "                     /protein_id=\"YP_005098101.1\"\n",
      "                     /db_xref=\"GeneID:11669863\"\n",
      "                     /translation=\"MWRESHPETVAFWRDLKDGILAAVRNPGREFWAGAHLRRNGERA\n",
      "                     IRIWRTVEFDSSGRKVPGWWLCMELPSGRILSYPGIGVSVTKETDEDGRINTNVRIKY\n",
      "                     QGENQLTRQWTTLYTHGGKACENIVQALCRDLLAYAMLNVEAGGYPIVLSVHDELVCE\n",
      "                     TPDTSDYTVAELEKLMCALPEWADGFPLVAEGAELKRYAK\"\n",
      "     gene            complement(3328..5712)\n",
      "                     /locus_tag=\"SaPhSE2_gp06\"\n",
      "                     /db_xref=\"GeneID:11669864\"\n",
      "     CDS             complement(3328..5712)\n",
      "                     /locus_tag=\"SaPhSE2_gp06\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=11\n",
      "                     /product=\"DNA polymerase\"\n",
      "                     /protein_id=\"YP_005098102.1\"\n",
      "                     /db_xref=\"GeneID:11669864\"\n",
      "                     /translation=\"MNLLYLDTETFSEADLKKVGSYAYAEHPTTEIVICTYAFDEGPV\n",
      "                     QVWDATDGSDMPRDLRRAMLKLQKPDSNLKLVGQNFLMFDRPVIKHCWGFELLVENII\n",
      "                     DTMIVAFRHALPGSLAALCEVLNIDASMAKDKRGKALIQRFSKPTPKNYKIRRYTADT\n",
      "                     HPKEWAEFIAYAKSDITSMREVYKKMPKWGNSEFEDRVLHLDQVINDRGFKVDVALAE\n",
      "                     AAIEAVTRHKEELQEEAQRKYGGSLTGKDFLPILQELAPAHRIHNAQKSTLNDLLADE\n",
      "                     DLPDDARTIIEMRLGAASTASTKYAPLLLGRSSDDRRRGCLQYGGAKRTLRWAGKGFQ\n",
      "                     PQNLARGYYHDDELDRGIAALLKGRAHRRFDVAKLTASTVRSCIIPEAGRKFVVADYS\n",
      "                     NVEGRGLAWLAGEETALDTFRAGLDIYCVTAGKMFGMDPDDIKKERKDLRQIGKACLH\n",
      "                     RHTQVLTDGGFKDIMAVTSAYKVWNGEKWVNTKGTHLMGWKPVINVDGVLMTEDHKIL\n",
      "                     THSWKAAKQLVSNKYMMGLALGRGGDAWLSYVSYQNDKAKDNYSSNVIAERCRGGSIM\n",
      "                     TMSERVKPLNATSAPLKQQKNIVKNISATKTQCRMMRTERDCSTGCPRQSCGQQAPET\n",
      "                     RVTKTMGLGVSRYLTSGGPIKGRFLDMFKLWKAGMTQTSKWTESTQTVTTSLATFGLS\n",
      "                     AVLKTAVTDVEYLSFNESMMQPLPALLSLNGKLTYCEPVYDLIDVEDGNRFLIASDSG\n",
      "                     FLVAHNCELGLGYEGGVGAFVTFAKNLGLDLIEMAKTMDGTFPRPHLGCYCTWV\"\n",
      "     gene            complement(5770..6396)\n",
      "                     /locus_tag=\"SaPhSE2_gp07\"\n",
      "                     /db_xref=\"GeneID:11669865\"\n",
      "     CDS             complement(5770..6396)\n",
      "                     /locus_tag=\"SaPhSE2_gp07\"\n",
      "                     /codon_start=1\n",
      "                     /transl_table=11\n",
      "                     /product=\"hypothetical protein\"\n",
      "                     /protein_id=\"YP_005098103.1\"\n",
      "                     /db_xref=\"GeneID:11669865\"\n",
      "                     /translation=\"MGIKLNLRKVQTAWLNVFERAKDRENNDGSVTKGTYNGTFILTP\n",
      "                     EHPQIEELRDTVFTVVSEALGEAAAEKWMKQNYGEGKHMDKCAVRDIAERDNPFEDFP\n",
      "                     EGFYFQAKNKQQPLILTSVKGEKQVEPDFNIDGEQIE Error: External viewer error: Empty Response. Bytes read: 0 Status: Timeout\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Extract genome sequences from the reports, add them to viral_genomes\n",
    "viral_genomes['Sequence'] = 'None'\n",
    "\n",
    "def get_genome(report):\n",
    "    try:\n",
    "        report_origin = re.findall('ORIGIN[0-9,a-z, ,\\n]+', report)\n",
    "        report_origin = report_origin[0]\n",
    "        origin = re.findall('[atcg]+', report_origin)\n",
    "        s = ''\n",
    "        genome = s.join(origin)\n",
    "        return(genome)\n",
    "    except IndexError:\n",
    "        print('error at', report)\n",
    "        pass\n",
    "\n",
    "\n",
    "viral_genomes['Sequence'] = meta['Report'].apply(get_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 9458 entries, 0 to 9457\n",
      "Data columns (total 13 columns):\n",
      "Genome                9458 non-null object\n",
      "Accession             9458 non-null object\n",
      "Source information    9458 non-null object\n",
      "Segm                  9458 non-null object\n",
      "Length                9458 non-null object\n",
      "Protein               9458 non-null object\n",
      "Neighbors             9458 non-null object\n",
      "Host                  9458 non-null object\n",
      "Created               9458 non-null object\n",
      "Updated               9458 non-null object\n",
      "Classification        9458 non-null object\n",
      "Extra Info            2996 non-null object\n",
      "Sequence              9457 non-null object\n",
      "dtypes: object(13)\n",
      "memory usage: 960.7+ KB\n"
     ]
    }
   ],
   "source": [
    "viral_genomes.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "sal_phage = viral_genomes.loc[viral_genomes['Sequence'].isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "##### for some reason the Salmonella phage SE2, complete genome report loaded in without a genome, so I'll try getting that again. \n",
    "Entrez.email = 'qdupupet@umass.edu'\n",
    "handle = Entrez.efetch(db=\"nuccore\", id='NC_016763.1', rettype=\"gb\", retmode=\"text\")\n",
    "sal_phage_2_report = handle.read()\n",
    "\n",
    "report_origin = re.findall('ORIGIN[0-9,a-z, ,\\n]+', sal_phage_2_report)\n",
    "report_origin = report_origin[0]\n",
    "origin = re.findall('[atcg]+', report_origin)\n",
    "s = ''\n",
    "sal_phage_2_genome = s.join(origin)\n",
    "\n",
    "\n",
    "\n",
    "sal_phage['Sequence'] = sal_phage['Sequence'].apply(lambda x: sal_phage_2_genome)\n",
    "\n",
    "viral_genomes[2879:2880] = sal_phage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 9458 entries, 0 to 9457\n",
      "Data columns (total 13 columns):\n",
      "Genome                9458 non-null object\n",
      "Accession             9458 non-null object\n",
      "Source information    9458 non-null object\n",
      "Segm                  9458 non-null object\n",
      "Length                9458 non-null object\n",
      "Protein               9458 non-null object\n",
      "Neighbors             9458 non-null object\n",
      "Host                  9458 non-null object\n",
      "Created               9458 non-null object\n",
      "Updated               9458 non-null object\n",
      "Classification        9458 non-null object\n",
      "Extra Info            2996 non-null object\n",
      "Sequence              9458 non-null object\n",
      "dtypes: object(13)\n",
      "memory usage: 960.7+ KB\n"
     ]
    }
   ],
   "source": [
    "viral_genomes.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Extract the molecule type (RNA or DNA) from the report\n",
    "viral_genomes['Molecule Type'] = 'None'\n",
    "\n",
    "def extract_genome_type(report):\n",
    "    try:\n",
    "        report_mol_type = re.findall('mol_type=[A-Z,a-z,=,\" ]+', report)\n",
    "        report_mol_type = report_mol_type[0]\n",
    "        molecule_type = re.findall('\"[A-z ]+\"', report_mol_type)\n",
    "        molecule_type = molecule_type[0].strip('\"')\n",
    "        return(molecule_type)\n",
    "    except IndexError:\n",
    "        print('error at', report)\n",
    "        pass\n",
    "    \n",
    "viral_genomes['Molecule Type'] = meta['Report'].apply(extract_genome_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Extract the phylogeny from the report\n",
    "viral_genomes['Phylogeny'] = 'None'\n",
    "\n",
    "def extract_phylogeny(report):\n",
    "    try:\n",
    "        report_phylo = re.findall('Viruses[A-z ;,]+', report)\n",
    "        report_phylo = report_phylo[0]\n",
    "        return(report_phylo)\n",
    "    except IndexError:\n",
    "        print('error at', report)\n",
    "        pass\n",
    "    \n",
    "viral_genomes['Phylogeny'] = meta['Report'].apply(extract_phylogeny)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 9458 entries, 0 to 9457\n",
      "Data columns (total 15 columns):\n",
      "Genome                9458 non-null object\n",
      "Accession             9458 non-null object\n",
      "Source information    9458 non-null object\n",
      "Segm                  9458 non-null object\n",
      "Length                9458 non-null object\n",
      "Protein               9458 non-null object\n",
      "Neighbors             9458 non-null object\n",
      "Host                  9458 non-null object\n",
      "Created               9458 non-null object\n",
      "Updated               9458 non-null object\n",
      "Classification        9458 non-null object\n",
      "Extra Info            2996 non-null object\n",
      "Sequence              9458 non-null object\n",
      "Molecule Type         9458 non-null object\n",
      "Phylogeny             9458 non-null object\n",
      "dtypes: object(15)\n",
      "memory usage: 1.1+ MB\n"
     ]
    }
   ],
   "source": [
    "viral_genomes.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "viral_genomes.to_csv('viral_genomes.csv', index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
