{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import bs4 as bs\n",
    "import requests\n",
    "import numpy as np\n",
    "import progressbar\n",
    "import os\n",
    "import csv\n",
    "from Bio import Entrez\n",
    "import re\n",
    "import time\n",
    "import sys"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Acquisition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Create a BeautifulSoup object for the main page\n",
    "url = 'https://www.ncbi.nlm.nih.gov/genomes/GenomesGroup.cgi'\n",
    "result = requests.get(url)\n",
    "main_soup = bs.BeautifulSoup(result.content, 'html.parser')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Parse the main_soup for the table of viruses\n",
    "table = main_soup.find('table', {'class': 'tblTxt'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# This table contains some subtables, which are included within single cells, called 'minitable' below\n",
    "# The rows in the minitables are formatted differently, but contain the same info that I want, so this function \n",
    "# will convert them into the same format as the rest. A list comprehension didn't work because the 'minirows' have \n",
    "# variable lengths\n",
    "def edit_minirow(row): \n",
    "    n = len(row)\n",
    "    #assuming edited rows should have structure =[row[0],row[2],'None','None', row[1],row[3], row[4]]\n",
    "    if n == 1:\n",
    "        new_minirow = [row[0], 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None', 'None']\n",
    "    elif n == 2:\n",
    "        new_minirow = [row[0], 'None', 'None', 'None', row[1], 'None', 'None', 'None', 'None', 'None', 'None']\n",
    "    elif n == 3:\n",
    "        new_minirow = [row[0], row[2], 'None', 'None', row[1], 'None', 'None', 'None', 'None', 'None', 'None']\n",
    "    elif n == 4:\n",
    "        new_minirow = [row[0], row[2], 'None', 'None', row[1], row[3], 'None', 'None', 'None', 'None', 'None']\n",
    "    elif n == 5:\n",
    "        new_minirow = [row[0], row[2], 'None', 'None', row[1], row[3], row[4], 'None', 'None', 'None', 'None']\n",
    "    else:\n",
    "        print('error:', row)\n",
    "    return(new_minirow)\n",
    "\n",
    "# Initialize the list of classifications\n",
    "class_list = ['Deltavirus', 'Genomoviridae', 'Retro-transcribing viruses', 'Satellites', \n",
    "'Virus families not assigned to an order', 'Virus-associated RNAs', 'dsDNA viruses, no RNA stage', 'dsRNA viruses',\n",
    "'environmental samples', 'ssDNA viruses', 'ssRNA viruses', 'unassigned viruses', 'unclassified RNA viruses',\n",
    "'unclassified archaeal viruses', 'unclassified bacterial viruses', 'unclassified virophages', \n",
    "'unclassified viruses']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "# Create an empty list of lists\n",
    "viral_genome_array = []\n",
    "\n",
    "# Create a list of rows that will ensure that for every row in a minitable, that row isn't copied again when the \n",
    "#  main table finds those rows. The alternative would be to not include a separate minitable code, but since the \n",
    "#  minitables are within cells, those cells' text is added to the table\n",
    "minirow_master_list = []\n",
    "\n",
    "# This is in order to add the previous non-minitable row to each minitable row, as it contains relevant info\n",
    "head_row = []\n",
    "\n",
    "# This variable will hold the subheader of the previous lighter colored column, which will contain extra info for the \n",
    "# light columns beneath it\n",
    "sub_head = ''\n",
    "sub_head_list = []\n",
    "\n",
    "# Go through all of the rows in the table \n",
    "for row in table.find_all('tr'):\n",
    "    this_row = []\n",
    "    # Go through each cell in each row\n",
    "    for cell in row.find_all('td'):\n",
    "        \n",
    "        # If the cell contains a table (the pulldown rows on the main page) reformat those rows like the normal rows\n",
    "        try:\n",
    "            minitable = cell.find('table')\n",
    "            for minirow in minitable.find_all('tr'):\n",
    "                this_minirow = []\n",
    "                for minicell in minirow.find_all('td'):\n",
    "                    if 'NC_' in minicell.text:\n",
    "                        this_minirow.append(minicell.find('a').attrs['href'])\n",
    "                    else:\n",
    "                        this_minirow.append(minicell.text.replace('\\n','').replace('\\xa0',''))\n",
    "                minirow_master_list.append(this_minirow)\n",
    "                edited_minirow = edit_minirow(this_minirow)\n",
    "                edited_minirow.append(head_row)\n",
    "                edited_minirow[0] = edited_minirow[0] + ' segment '\n",
    "                viral_genome_array.append(edited_minirow)\n",
    "                \n",
    "        # If there isn't a table, grab the href link for cells with a 'NC_', otherwise grab the text\n",
    "        except AttributeError:\n",
    "            if 'NC_' in cell.text:\n",
    "                this_row.append(cell.find('a').attrs['href'])\n",
    "            else:\n",
    "                this_row.append(cell.text.replace('\\n','').replace('\\xa0',''))\n",
    "    \n",
    "    # For the lighter rows, which share a subheader, either grab the subheader or append the current subheader in \n",
    "    # the 12th position (index 11)\n",
    "    try:\n",
    "        if row.attrs['bgcolor'] == '#F8F8F8':\n",
    "            if len(this_row) == 1:\n",
    "                # Classifications also have this lighter color, but I will to treat them differently to add them to \n",
    "                #  every row\n",
    "                if this_row[0] not in class_list:\n",
    "                    sub_head = this_row[0]\n",
    "                    sub_head_list.append(this_row)\n",
    "            else:\n",
    "                while len(this_row) < 11:\n",
    "                    this_row.append('None')\n",
    "                this_row.append(sub_head)\n",
    "    except:\n",
    "        pass\n",
    "    #\n",
    "    \n",
    "    # don't take rows that have already been taken down and reformatteed, and rows that include the text \n",
    "    # of a whole minitable (which have the name repeated)\n",
    "    if this_row not in minirow_master_list and this_row[0] not in this_row[1:]: \n",
    "        head_row = this_row \n",
    "        viral_genome_array.append(this_row)\n",
    "    \n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Add a classification and extra info feature in the first row (these will become dataframe column headers)\n",
    "viral_genome_array[0].append('Classification')\n",
    "viral_genome_array[0].append('Extra Info')\n",
    "\n",
    "# If a row contains only a classification, everything below that row falls into that category until the next \n",
    "# classification comes up\n",
    "classification = 'None'\n",
    "for row in viral_genome_array[1:]:\n",
    "    if row[0] in class_list:\n",
    "        classification = row[0]\n",
    "        viral_genome_array.remove(row)\n",
    "    else:\n",
    "        while len(row) < 11:\n",
    "            row.append('None')\n",
    "        row[10] = classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Gather a list of the head_rows that ended up in the extra info columns\n",
    "head_row_list = []\n",
    "for row in viral_genome_array[1:]:\n",
    "    if len(row) == 12:\n",
    "        head = row[11]\n",
    "        if head != 'None' and head not in head_row_list:\n",
    "            head_row_list.append(head)\n",
    "\n",
    "# now that I have a list of heads, go through the array again and remove rows that match those head_rows, as well\n",
    "#  as subheaders\n",
    "for row in viral_genome_array[1:]:\n",
    "    if row in head_row_list or row in sub_head_list:\n",
    "        viral_genome_array.remove(row)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Extract the info in 'Extra Info' lists \n",
    "# For each row in the array, if the row ends in a list, replace missing values in that row with the corresponding \n",
    "#  value in the list\n",
    "for row in viral_genome_array:\n",
    "    if type(row[-1]) == list:\n",
    "        extra_list = row[-1]\n",
    "        for i in range(len(row)):\n",
    "            if i not in [3,4,5,6]:\n",
    "                if row[i] == 'None' or row[i] == '' or row[i] == '-':\n",
    "                    row[i] = extra_list[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Number all of the segments\n",
    "segment_counts = {}\n",
    "\n",
    "for row in viral_genome_array:\n",
    "    name = row[0]\n",
    "    if 'segment' in name:\n",
    "        if name not in segment_counts:\n",
    "            segment_counts[name] = 1\n",
    "            row[0] += '1'\n",
    "        else:\n",
    "            segment_counts[name] += 1\n",
    "            row[0] += str(segment_counts[name])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Source information</th>\n",
       "      <th>Segm</th>\n",
       "      <th>Length</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Neighbors</th>\n",
       "      <th>Host</th>\n",
       "      <th>Created</th>\n",
       "      <th>Updated</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Extra Info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hepatitis delta virus</td>\n",
       "      <td>/nuccore/13277517</td>\n",
       "      <td></td>\n",
       "      <td>-</td>\n",
       "      <td>1682nt</td>\n",
       "      <td>2</td>\n",
       "      <td>320</td>\n",
       "      <td>vertebrates, human</td>\n",
       "      <td>04/29/1993</td>\n",
       "      <td>02/10/2015</td>\n",
       "      <td>Deltavirus</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Badger associated gemykibivirus 1</td>\n",
       "      <td>/nuccore/807743872</td>\n",
       "      <td>strain:588t</td>\n",
       "      <td>-</td>\n",
       "      <td>2112nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>04/16/2015</td>\n",
       "      <td>11/24/2015</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bemisia-associated genomovirus AdDF</td>\n",
       "      <td>/nuccore/1211677462</td>\n",
       "      <td>isolate:AdDF</td>\n",
       "      <td>-</td>\n",
       "      <td>2199nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/19/2017</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bemisia-associated genomovirus AdO</td>\n",
       "      <td>/nuccore/1211677465</td>\n",
       "      <td>isolate:AdO</td>\n",
       "      <td>-</td>\n",
       "      <td>2211nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/19/2017</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bemisia-associated genomovirus NfO</td>\n",
       "      <td>/nuccore/1211677468</td>\n",
       "      <td>isolate:NfO</td>\n",
       "      <td>-</td>\n",
       "      <td>2231nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/19/2017</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Genome            Accession  \\\n",
       "0                Hepatitis delta virus    /nuccore/13277517   \n",
       "1    Badger associated gemykibivirus 1   /nuccore/807743872   \n",
       "2  Bemisia-associated genomovirus AdDF  /nuccore/1211677462   \n",
       "3   Bemisia-associated genomovirus AdO  /nuccore/1211677465   \n",
       "4   Bemisia-associated genomovirus NfO  /nuccore/1211677468   \n",
       "\n",
       "  Source information Segm  Length Protein Neighbors                Host  \\\n",
       "0                       -  1682nt       2       320  vertebrates, human   \n",
       "1        strain:588t    -  2112nt       2         -         vertebrates   \n",
       "2       isolate:AdDF    -  2199nt       2         -       invertebrates   \n",
       "3        isolate:AdO    -  2211nt       2         -       invertebrates   \n",
       "4        isolate:NfO    -  2231nt       2         -       invertebrates   \n",
       "\n",
       "      Created     Updated Classification Extra Info  \n",
       "0  04/29/1993  02/10/2015     Deltavirus       None  \n",
       "1  04/16/2015  11/24/2015  Genomoviridae       None  \n",
       "2  06/27/2017  07/19/2017  Genomoviridae       None  \n",
       "3  06/27/2017  07/19/2017  Genomoviridae       None  \n",
       "4  06/27/2017  07/19/2017  Genomoviridae       None  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "viral_genomes = pd.DataFrame(viral_genome_array[2:], columns = viral_genome_array[0])#, columns = viral_genome_array[0])\n",
    "viral_genomes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#------------------------------------------------------------------------------------------------------------------\n",
    "##### This cell gets the full html 'soup' from each virus' page, which includes the genome. My intention is to only run this once since it will take a while, and save the result into a csv file in this directory. \n",
    "length = len(viral_genomes['Genome'])\n",
    "\n",
    "def scrape_viruses(start = 0, end = 9460):\n",
    "    soup_dict = {}\n",
    "    ###### Using a progress bar, go through the dataframe, getting a soup for each genome and saving the result to a dict\n",
    "    with progressbar.ProgressBar(max_value=length) as bar:\n",
    "        for i in range(start, end):\n",
    "\n",
    "            name = viral_genomes['Genome'][i]\n",
    "\n",
    "            slug = viral_genomes['Accession'][i]\n",
    "\n",
    "            result = requests.get('https://www.ncbi.nlm.nih.gov{}'.format(slug))\n",
    "\n",
    "            soup = bs.BeautifulSoup(result.content, 'html.parser')\n",
    "\n",
    "            soup_dict[name] = soup\n",
    "\n",
    "            bar.update(i)\n",
    "\n",
    "    os.system('say \"download complete\"')\n",
    "\n",
    "    ###### Write the dict to a csv: \"virus_soups.csv\"\n",
    "    w = csv.writer(open(\"virus_soups_{}.csv\".format(str(start) + '-' + str(end)), \"w\"))\n",
    "    for key, val in soup_dict.items():\n",
    "        w.writerow([key, val])\n",
    "    os.system('say \"file saved\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    scrape_viruses(0, 1000)\n",
    "\n",
    "    scrape_viruses(1000,2000)\n",
    "\n",
    "    scrape_viruses(2000,3000)\n",
    "\n",
    "    scrape_viruses(3000,4000)\n",
    "\n",
    "    scrape_viruses(4000,5000)\n",
    "\n",
    "    scrape_viruses(5000,6000)\n",
    "\n",
    "    scrape_viruses(6000,7000)\n",
    "\n",
    "    scrape_viruses(7000,8000)\n",
    "\n",
    "    scrape_viruses(start = 8000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "soup0 = pd.read_csv('virus_soups_0-1000.csv', header = None, names = ['Name', 'BeautifulSoup'])\n",
    "soup1 = pd.read_csv('virus_soups_1000-2000.csv', header = None, names = ['Name', 'BeautifulSoup'])\n",
    "soup2 = pd.read_csv('virus_soups_2000-3000.csv', header = None, names = ['Name', 'BeautifulSoup'])\n",
    "soup3 = pd.read_csv('virus_soups_3000-4000.csv', header = None, names = ['Name', 'BeautifulSoup'])\n",
    "soup4 = pd.read_csv('virus_soups_4000-5000.csv', header = None, names = ['Name', 'BeautifulSoup'])\n",
    "soup5 = pd.read_csv('virus_soups_5000-6000.csv', header = None, names = ['Name', 'BeautifulSoup'])\n",
    "soup6 = pd.read_csv('virus_soups_6000-7000.csv', header = None, names = ['Name', 'BeautifulSoup'])\n",
    "soup7 = pd.read_csv('virus_soups_7000-8000.csv', header = None, names = ['Name', 'BeautifulSoup'])\n",
    "soup8 = pd.read_csv('virus_soups_8000-9460.csv', header = None, names = ['Name', 'BeautifulSoup'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "soups = pd.concat([soup0, soup1, soup2, soup3, soup4, soup5, soup6, soup7, soup8], ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Source information</th>\n",
       "      <th>Segm</th>\n",
       "      <th>Length</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Neighbors</th>\n",
       "      <th>Host</th>\n",
       "      <th>Created</th>\n",
       "      <th>Updated</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Extra Info</th>\n",
       "      <th>BeautifulSoup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hepatitis delta virus</td>\n",
       "      <td>/nuccore/13277517</td>\n",
       "      <td></td>\n",
       "      <td>-</td>\n",
       "      <td>1682nt</td>\n",
       "      <td>2</td>\n",
       "      <td>320</td>\n",
       "      <td>vertebrates, human</td>\n",
       "      <td>04/29/1993</td>\n",
       "      <td>02/10/2015</td>\n",
       "      <td>Deltavirus</td>\n",
       "      <td>None</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Badger associated gemykibivirus 1</td>\n",
       "      <td>/nuccore/807743872</td>\n",
       "      <td>strain:588t</td>\n",
       "      <td>-</td>\n",
       "      <td>2112nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>04/16/2015</td>\n",
       "      <td>11/24/2015</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>None</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bemisia-associated genomovirus AdDF</td>\n",
       "      <td>/nuccore/1211677462</td>\n",
       "      <td>isolate:AdDF</td>\n",
       "      <td>-</td>\n",
       "      <td>2199nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/19/2017</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>None</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bemisia-associated genomovirus AdO</td>\n",
       "      <td>/nuccore/1211677465</td>\n",
       "      <td>isolate:AdO</td>\n",
       "      <td>-</td>\n",
       "      <td>2211nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/19/2017</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>None</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bemisia-associated genomovirus NfO</td>\n",
       "      <td>/nuccore/1211677468</td>\n",
       "      <td>isolate:NfO</td>\n",
       "      <td>-</td>\n",
       "      <td>2231nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/19/2017</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>None</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Genome            Accession  \\\n",
       "0                Hepatitis delta virus    /nuccore/13277517   \n",
       "1    Badger associated gemykibivirus 1   /nuccore/807743872   \n",
       "2  Bemisia-associated genomovirus AdDF  /nuccore/1211677462   \n",
       "3   Bemisia-associated genomovirus AdO  /nuccore/1211677465   \n",
       "4   Bemisia-associated genomovirus NfO  /nuccore/1211677468   \n",
       "\n",
       "  Source information Segm  Length Protein Neighbors                Host  \\\n",
       "0                       -  1682nt       2       320  vertebrates, human   \n",
       "1        strain:588t    -  2112nt       2         -         vertebrates   \n",
       "2       isolate:AdDF    -  2199nt       2         -       invertebrates   \n",
       "3        isolate:AdO    -  2211nt       2         -       invertebrates   \n",
       "4        isolate:NfO    -  2231nt       2         -       invertebrates   \n",
       "\n",
       "      Created     Updated Classification Extra Info  \\\n",
       "0  04/29/1993  02/10/2015     Deltavirus       None   \n",
       "1  04/16/2015  11/24/2015  Genomoviridae       None   \n",
       "2  06/27/2017  07/19/2017  Genomoviridae       None   \n",
       "3  06/27/2017  07/19/2017  Genomoviridae       None   \n",
       "4  06/27/2017  07/19/2017  Genomoviridae       None   \n",
       "\n",
       "                                       BeautifulSoup  \n",
       "0  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...  \n",
       "1  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...  \n",
       "2  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...  \n",
       "3  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...  \n",
       "4  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "viral_genomes['BeautifulSoup'] = soups['BeautifulSoup'].apply(lambda x: bs.BeautifulSoup(x, 'html.parser'))\n",
    "viral_genomes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 9460 entries, 0 to 9459\n",
      "Data columns (total 13 columns):\n",
      "Genome                9460 non-null object\n",
      "Accession             9460 non-null object\n",
      "Source information    9460 non-null object\n",
      "Segm                  9460 non-null object\n",
      "Length                9460 non-null object\n",
      "Protein               9460 non-null object\n",
      "Neighbors             9460 non-null object\n",
      "Host                  9460 non-null object\n",
      "Created               9460 non-null object\n",
      "Updated               9460 non-null object\n",
      "Classification        9460 non-null object\n",
      "Extra Info            3000 non-null object\n",
      "BeautifulSoup         9460 non-null object\n",
      "dtypes: object(13)\n",
      "memory usage: 960.9+ KB\n"
     ]
    }
   ],
   "source": [
    "viral_genomes.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# This function returns the id that I need for each virus to access it in the NCBI Entrez database \n",
    "def get_db_id (soup):\n",
    "    try:\n",
    "        line = soup.find('p', {'class': 'itemid'}).text\n",
    "        itemid = (line[25:])\n",
    "        return(itemid)\n",
    "    except AttributeError:\n",
    "        return(np.nan)\n",
    "\n",
    "# Add the sequence type to the main DataFrame\n",
    "def get_seq_type(soup):\n",
    "    try:\n",
    "        head = soup.find('div', {\"class\":\"rprtheader\"})\n",
    "        line = head.find('h1')\n",
    "        title = (line.text)\n",
    "        sequence_type = re.search(',[a-z,0-9, ]+', title)[0]\n",
    "        return(sequence_type[2:])\n",
    "    except AttributeError:\n",
    "        return('error')\n",
    "    except TypeError:\n",
    "        return('error')    \n",
    "    \n",
    "viral_genomes['ItemId'] = viral_genomes['BeautifulSoup'].apply(get_db_id)\n",
    "viral_genomes['SequenceType'] = viral_genomes['BeautifulSoup'].apply(get_seq_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 9460 entries, 0 to 9459\n",
      "Data columns (total 15 columns):\n",
      "Genome                9460 non-null object\n",
      "Accession             9460 non-null object\n",
      "Source information    9460 non-null object\n",
      "Segm                  9460 non-null object\n",
      "Length                9460 non-null object\n",
      "Protein               9460 non-null object\n",
      "Neighbors             9460 non-null object\n",
      "Host                  9460 non-null object\n",
      "Created               9460 non-null object\n",
      "Updated               9460 non-null object\n",
      "Classification        9460 non-null object\n",
      "Extra Info            3000 non-null object\n",
      "BeautifulSoup         9460 non-null object\n",
      "ItemId                9458 non-null object\n",
      "SequenceType          9460 non-null object\n",
      "dtypes: object(15)\n",
      "memory usage: 1.1+ MB\n"
     ]
    }
   ],
   "source": [
    "# Check to make sure we have an itemid for each virus/segment\n",
    "viral_genomes.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "624\n",
      "625\n"
     ]
    }
   ],
   "source": [
    "# identify missing ids (np.nan is a float type)\n",
    "for i in range(len(viral_genomes['ItemId'])):\n",
    "    string = viral_genomes['ItemId'][i]\n",
    "    if type(string) == float:\n",
    "        print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Source information</th>\n",
       "      <th>Segm</th>\n",
       "      <th>Length</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Neighbors</th>\n",
       "      <th>Host</th>\n",
       "      <th>Created</th>\n",
       "      <th>Updated</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Extra Info</th>\n",
       "      <th>BeautifulSoup</th>\n",
       "      <th>ItemId</th>\n",
       "      <th>SequenceType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>624</th>\n",
       "      <td>Bacillus phage Bobb</td>\n",
       "      <td>/nuccore/682123955</td>\n",
       "      <td></td>\n",
       "      <td>-</td>\n",
       "      <td>160281nt</td>\n",
       "      <td>247</td>\n",
       "      <td>-</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>09/09/2014</td>\n",
       "      <td>09/09/2014</td>\n",
       "      <td>dsDNA viruses, no RNA stage</td>\n",
       "      <td>None</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>error</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>625</th>\n",
       "      <td>Bacillus phage Bp8p-C</td>\n",
       "      <td>/nuccore/985761324</td>\n",
       "      <td></td>\n",
       "      <td>-</td>\n",
       "      <td>151417nt</td>\n",
       "      <td>211</td>\n",
       "      <td>-</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>02/05/2016</td>\n",
       "      <td>02/05/2016</td>\n",
       "      <td>dsDNA viruses, no RNA stage</td>\n",
       "      <td>None</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>error</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Genome           Accession Source information Segm  \\\n",
       "624    Bacillus phage Bobb  /nuccore/682123955                       -   \n",
       "625  Bacillus phage Bp8p-C  /nuccore/985761324                       -   \n",
       "\n",
       "       Length Protein Neighbors      Host     Created     Updated  \\\n",
       "624  160281nt     247         -  bacteria  09/09/2014  09/09/2014   \n",
       "625  151417nt     211         -  bacteria  02/05/2016  02/05/2016   \n",
       "\n",
       "                  Classification Extra Info  \\\n",
       "624  dsDNA viruses, no RNA stage       None   \n",
       "625  dsDNA viruses, no RNA stage       None   \n",
       "\n",
       "                                         BeautifulSoup ItemId SequenceType  \n",
       "624  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...    NaN        error  \n",
       "625  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...    NaN        error  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "viral_genomes[624:626]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#manually enter ids for 624, 625\n",
    "\n",
    "viral_genomes['ItemId'][624] = 'NC_024792.1'\n",
    "viral_genomes['ItemId'][625] = 'NC_029121'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 9460 entries, 0 to 9459\n",
      "Data columns (total 15 columns):\n",
      "Genome                9460 non-null object\n",
      "Accession             9460 non-null object\n",
      "Source information    9460 non-null object\n",
      "Segm                  9460 non-null object\n",
      "Length                9460 non-null object\n",
      "Protein               9460 non-null object\n",
      "Neighbors             9460 non-null object\n",
      "Host                  9460 non-null object\n",
      "Created               9460 non-null object\n",
      "Updated               9460 non-null object\n",
      "Classification        9460 non-null object\n",
      "Extra Info            3000 non-null object\n",
      "BeautifulSoup         9460 non-null object\n",
      "ItemId                9460 non-null object\n",
      "SequenceType          9460 non-null object\n",
      "dtypes: object(15)\n",
      "memory usage: 1.1+ MB\n"
     ]
    }
   ],
   "source": [
    "viral_genomes.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "complete genome                                                                                                                                                     5528\n",
       "complete sequence                                                                                                                                                   1683\n",
       "complete cds                                                                                                                                                        1414\n",
       "error                                                                                                                                                                202\n",
       "hypothetical protein 2, and hypothetical protein 3 genes, complete cds                                                                                                93\n",
       "                                                                                                                                                                      63\n",
       "segment                                                                                                                                                               54\n",
       "isolate                                                                                                                                                               35\n",
       "complete genome, isolate                                                                                                                                              33\n",
       "partial genome                                                                                                                                                        31\n",
       "hypothetical protein 2, hypothetical protein 3, and hypothetical protein 4 genes, complete cds                                                                        27\n",
       "complete genome, strain                                                                                                                                               16\n",
       "genome                                                                                                                                                                11\n",
       "genomic                                                                                                                                                               11\n",
       "strain                                                                                                                                                                 9\n",
       "complete genome, isolate                                                                                                                                               8\n",
       "complete cds, strain                                                                                                                                                   8\n",
       "putative phosphoprotein                                                                                                                                                7\n",
       "hypothetical protein, putative structural protein, and hypothetical protein genes, complete cds                                                                        7\n",
       "partial cds                                                                                                                                                            7\n",
       "putative glycoprotein, and hypothetical protein genes, complete cds                                                                                                    6\n",
       "gene, complete cds                                                                                                                                                     6\n",
       "phosphoprotein, hypothetical proteins, matrix, glycoprotein, hypothetical protein, and polymerase genes, complete cds                                                  6\n",
       "hypothetical protein 2, hypothetical protein 3, putative glycoprotein, and                                                                                             6\n",
       "complete cds, isolate                                                                                                                                                  5\n",
       "phosphoprotein                                                                                                                                                         5\n",
       "phosphoprotein, matrix, glycoprotein, and polymerase genes, complete cds                                                                                               5\n",
       "hypothetical protein, and                                                                                                                                              5\n",
       "hypothetical protein 2, hypothetical protein 3, hypothetical protein 4, and hypothetical protein 5 genes, complete cds                                                 4\n",
       "hypothetical protein 1, and hypothetical protein 2 genes, complete cds                                                                                                 4\n",
       "                                                                                                                                                                    ... \n",
       "capsid protein, and protein delta genes, complete cds                                                                                                                  1\n",
       "hypothetical protein 2, putative                                                                                                                                       1\n",
       "nearly complete genome, strain                                                                                                                                         1\n",
       "putative structural protein, and hypothetical protein genes, complete cds                                                                                              1\n",
       "hypothetical protein, putative capsid protein, and hypothetical protein genes, complete cds                                                                            1\n",
       "phosphoprotein, matrix, hypothetical proteins, glycoprotein, hypothetical protein, and polymerase genes, complete cds                                                  1\n",
       "capsid protein and movement protein                                                                                                                                    1\n",
       "replication enhancer protein                                                                                                                                           1\n",
       "isolate w4                                                                                                                                                             1\n",
       "segment 1, genomic                                                                                                                                                     1\n",
       "triple gene block protein 1                                                                                                                                            1\n",
       "genes, complete cds                                                                                                                                                    1\n",
       "hypothetical protein, phosphoprotein, hypothetical protein, matrix, hypothetical protein, glycoprotein, hypothetical protein, and polymerase genes, complete cds       1\n",
       "hypothetical proteins, glycoprotein, and putative                                                                                                                      1\n",
       "hypothetical protein, putative                                                                                                                                         1\n",
       "membrane                                                                                                                                                               1\n",
       "isolate 1                                                                                                                                                              1\n",
       "segment 2, genomic                                                                                                                                                     1\n",
       "triple gene block protein 1, triple gene block protein 2, triple gene block protein 3, coat protein, and putative 10 k                                                 1\n",
       "putative                                                                                                                                                               1\n",
       "nonstructural protein 1                                                                                                                                                1\n",
       "spike glycoprotein, envelope protein, membrane protein, and nucleocapsid protein genes, complete cds                                                                   1\n",
       "hypothetical protein 2, putative membrane protein, putative glycoprotein 1, putative glycoprotein 2, and                                                               1\n",
       "hypothetical protein, putative matrix protein, putative glycoprotein, and                                                                                              1\n",
       "putative readthrough protein, and coat protein genes, complete cds                                                                                                     1\n",
       "6                                                                                                                                                                      1\n",
       "isolate 3                                                                                                                                                              1\n",
       "nearly complete genome                                                                                                                                                 1\n",
       "hypothetical protein 2, hypothetical protein 3, and putative glycoprotein genes, complete cds                                                                          1\n",
       "complete segment 2, genomic                                                                                                                                            1\n",
       "Name: SequenceType, Length: 151, dtype: int64"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "viral_genomes['SequenceType'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "viral_genomes.to_csv('viral_genomes_soups.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get NCBI reports with Entrez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "###### As with scraping the html for each virus above, I will attempt to only extract the report for each virus once, \n",
    "###### I will save this to a df and thereafter I'll be importing that and not running this cell. \n",
    "\n",
    "length = len(viral_genomes['Genome'])\n",
    "\n",
    "def get_virus_reports(start = 0, end = length):\n",
    "    report_dict = {}\n",
    "    with progressbar.ProgressBar(max_value=length) as bar:\n",
    "        for i in range(start, end):     #starting at 4235 because it stopped at 4240\n",
    "            \n",
    "            name = viral_genomes['Genome'][i]\n",
    "            try:\n",
    "                Entrez.email = 'qdupupet@umass.edu'\n",
    "                handle = Entrez.efetch(db=\"nuccore\", id=viral_genomes['ItemId'][i], rettype=\"gb\", retmode=\"text\")\n",
    "                report = handle.read()\n",
    "                report_dict[name] = report\n",
    "            except AttributeErrror:\n",
    "                pass\n",
    "            time.sleep(0.5)\n",
    "            bar.update(i)\n",
    "\n",
    "    os.system('say \"download complete\"')\n",
    "    \n",
    "    ###### Write the dict to a csv: \"virus_soups.csv\"\n",
    "    w = csv.writer(open(\"virus_reports_{}.csv\".format(str(start) + '-' + str(end)), \"w\"))\n",
    "    for key, val in report_dict.items():\n",
    "        w.writerow([key, val])\n",
    "    os.system('say \"file saved\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    get_virus_reports(0,2000)\n",
    "\n",
    "    get_virus_reports(2000,4000)\n",
    "\n",
    "    get_virus_reports(4000,6000)\n",
    "\n",
    "    get_virus_reports(6000,8000)\n",
    "\n",
    "    get_virus_reports(8000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "report0 = pd.read_csv('virus_reports_0-2000.csv', header = None, names = ['Name', 'Report'])\n",
    "report1 = pd.read_csv('virus_reports_2000-4000.csv', header = None, names = ['Name', 'Report'])\n",
    "report2 = pd.read_csv('virus_reports_4000-6000.csv', header = None, names = ['Name', 'Report'])\n",
    "report3 = pd.read_csv('virus_reports_6000-8000.csv', header = None, names = ['Name', 'Report'])\n",
    "report4 = pd.read_csv('virus_reports_8000-9460.csv', header = None, names = ['Name', 'Report'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "reports = pd.concat([report0, report1, report2, report3, report4], ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "viral_genomes['Report'] = reports['Report']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "viral_genomes.to_csv('viral_genomes_soups_and_reports')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extract genome sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "## read in the full meta df with all reports.\n",
    "#chunksize = 1000\n",
    "#TextFileReader = pd.read_csv('viral_genomes_soups_and_reports.csv', chunksize=chunksize, iterator=True)\n",
    "#viral_genomes = pd.concat(TextFileReader, ignore_index=True)\n",
    "\n",
    "#viral_genomes.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Extract genome sequences from the reports, add them to viral_genomes\n",
    "viral_genomes['Sequence'] = 'None'\n",
    "\n",
    "def get_genome(report):\n",
    "    \n",
    "    try:\n",
    "        report_origin = re.findall('ORIGIN[0-9,a-z, ,\\n]+', report)\n",
    "        report_origin = report_origin[0]\n",
    "        origin = re.findall('[atcg]+', report_origin)\n",
    "        s = ''\n",
    "        genome = s.join(origin)\n",
    "        return(genome)\n",
    "    except IndexError:\n",
    "        print('error at', report)\n",
    "        pass\n",
    "\n",
    "\n",
    "viral_genomes['Sequence'] = viral_genomes['Report'].apply(get_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 9460 entries, 0 to 9459\n",
      "Data columns (total 17 columns):\n",
      "Genome                9460 non-null object\n",
      "Accession             9460 non-null object\n",
      "Source information    9460 non-null object\n",
      "Segm                  9460 non-null object\n",
      "Length                9460 non-null object\n",
      "Protein               9460 non-null object\n",
      "Neighbors             9460 non-null object\n",
      "Host                  9460 non-null object\n",
      "Created               9460 non-null object\n",
      "Updated               9460 non-null object\n",
      "Classification        9460 non-null object\n",
      "Extra Info            3000 non-null object\n",
      "BeautifulSoup         9460 non-null object\n",
      "ItemId                9460 non-null object\n",
      "SequenceType          9460 non-null object\n",
      "Report                9460 non-null object\n",
      "Sequence              9460 non-null object\n",
      "dtypes: object(17)\n",
      "memory usage: 1.2+ MB\n"
     ]
    }
   ],
   "source": [
    "viral_genomes.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Source information</th>\n",
       "      <th>Segm</th>\n",
       "      <th>Length</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Neighbors</th>\n",
       "      <th>Host</th>\n",
       "      <th>Created</th>\n",
       "      <th>Updated</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Extra Info</th>\n",
       "      <th>BeautifulSoup</th>\n",
       "      <th>ItemId</th>\n",
       "      <th>SequenceType</th>\n",
       "      <th>Report</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9455</th>\n",
       "      <td>Wilkie partiti-like virus 1</td>\n",
       "      <td>/nuccore/1211677385</td>\n",
       "      <td>strain:mosWSCP36002</td>\n",
       "      <td>-</td>\n",
       "      <td>2100nt</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/14/2017</td>\n",
       "      <td>unclassified viruses</td>\n",
       "      <td>None</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "      <td>NC_035122.1</td>\n",
       "      <td>complete genome</td>\n",
       "      <td>LOCUS       NC_035122               2100 bp   ...</td>\n",
       "      <td>tatactacaatatatacaatatttttcactcataaaatgactgact...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9456</th>\n",
       "      <td>Wilkie partiti-like virus 2</td>\n",
       "      <td>/nuccore/1211677379</td>\n",
       "      <td>strain:mosWSCP53020</td>\n",
       "      <td>-</td>\n",
       "      <td>1819nt</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>eukaryotes</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/14/2017</td>\n",
       "      <td>unclassified viruses</td>\n",
       "      <td>None</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "      <td>NC_035119.1</td>\n",
       "      <td>complete genome</td>\n",
       "      <td>LOCUS       NC_035119               1819 bp   ...</td>\n",
       "      <td>ggcaatatgaattcctatgcacttaacaacattattcgtacttcac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9457</th>\n",
       "      <td>Wolkberg virus segment 1</td>\n",
       "      <td>/nuccore/1197509331</td>\n",
       "      <td>isolate:2562_SA3</td>\n",
       "      <td>None</td>\n",
       "      <td>(4461 nt)</td>\n",
       "      <td>proteins: 1</td>\n",
       "      <td>neighbors: 10</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>05/24/2017</td>\n",
       "      <td>05/24/2017</td>\n",
       "      <td>unclassified viruses</td>\n",
       "      <td>[Wolkberg virus, , isolate:2562_SA3, 3, 12312n...</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "      <td>NC_034631.1</td>\n",
       "      <td>complete cds</td>\n",
       "      <td>LOCUS       NC_034631               4461 bp   ...</td>\n",
       "      <td>agtagtgtactaccaatattctaatattttcaatatatctttacaa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9458</th>\n",
       "      <td>Wolkberg virus segment 2</td>\n",
       "      <td>/nuccore/1197509335</td>\n",
       "      <td>isolate:2562_SA3</td>\n",
       "      <td>None</td>\n",
       "      <td>(6873 nt)</td>\n",
       "      <td>proteins: 1</td>\n",
       "      <td>neighbors: 10</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>05/24/2017</td>\n",
       "      <td>05/24/2017</td>\n",
       "      <td>unclassified viruses</td>\n",
       "      <td>[Wolkberg virus, , isolate:2562_SA3, 3, 12312n...</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "      <td>NC_034633.1</td>\n",
       "      <td>complete cds</td>\n",
       "      <td>LOCUS       NC_034633               6873 bp   ...</td>\n",
       "      <td>agtagtgtactcctatactaacttcacttatcactaaaatggacga...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9459</th>\n",
       "      <td>Wolkberg virus segment 3</td>\n",
       "      <td>/nuccore/1197509333</td>\n",
       "      <td>isolate:2562_SA3</td>\n",
       "      <td>None</td>\n",
       "      <td>(978 nt)</td>\n",
       "      <td>proteins: 1</td>\n",
       "      <td>neighbors: 10</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>05/24/2017</td>\n",
       "      <td>05/24/2017</td>\n",
       "      <td>unclassified viruses</td>\n",
       "      <td>[Wolkberg virus, , isolate:2562_SA3, 3, 12312n...</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\n",
       "&lt;!DOCTY...</td>\n",
       "      <td>NC_034632.1</td>\n",
       "      <td>complete cds</td>\n",
       "      <td>LOCUS       NC_034632                978 bp   ...</td>\n",
       "      <td>agtagtgtactccagaaaaaagacaataacaaatctcaatctacaa...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Genome            Accession   Source information  \\\n",
       "9455  Wilkie partiti-like virus 1  /nuccore/1211677385  strain:mosWSCP36002   \n",
       "9456  Wilkie partiti-like virus 2  /nuccore/1211677379  strain:mosWSCP53020   \n",
       "9457     Wolkberg virus segment 1  /nuccore/1197509331     isolate:2562_SA3   \n",
       "9458     Wolkberg virus segment 2  /nuccore/1197509335     isolate:2562_SA3   \n",
       "9459     Wolkberg virus segment 3  /nuccore/1197509333     isolate:2562_SA3   \n",
       "\n",
       "      Segm     Length      Protein      Neighbors           Host     Created  \\\n",
       "9455     -     2100nt            1              -  invertebrates  06/27/2017   \n",
       "9456     -     1819nt            1              1     eukaryotes  06/27/2017   \n",
       "9457  None  (4461 nt)  proteins: 1  neighbors: 10  invertebrates  05/24/2017   \n",
       "9458  None  (6873 nt)  proteins: 1  neighbors: 10  invertebrates  05/24/2017   \n",
       "9459  None   (978 nt)  proteins: 1  neighbors: 10  invertebrates  05/24/2017   \n",
       "\n",
       "         Updated        Classification  \\\n",
       "9455  07/14/2017  unclassified viruses   \n",
       "9456  07/14/2017  unclassified viruses   \n",
       "9457  05/24/2017  unclassified viruses   \n",
       "9458  05/24/2017  unclassified viruses   \n",
       "9459  05/24/2017  unclassified viruses   \n",
       "\n",
       "                                             Extra Info  \\\n",
       "9455                                               None   \n",
       "9456                                               None   \n",
       "9457  [Wolkberg virus, , isolate:2562_SA3, 3, 12312n...   \n",
       "9458  [Wolkberg virus, , isolate:2562_SA3, 3, 12312n...   \n",
       "9459  [Wolkberg virus, , isolate:2562_SA3, 3, 12312n...   \n",
       "\n",
       "                                          BeautifulSoup       ItemId  \\\n",
       "9455  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...  NC_035122.1   \n",
       "9456  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...  NC_035119.1   \n",
       "9457  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...  NC_034631.1   \n",
       "9458  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...  NC_034633.1   \n",
       "9459  <?xml version=\"1.0\" encoding=\"utf-8\"?>\n",
       "<!DOCTY...  NC_034632.1   \n",
       "\n",
       "         SequenceType                                             Report  \\\n",
       "9455  complete genome  LOCUS       NC_035122               2100 bp   ...   \n",
       "9456  complete genome  LOCUS       NC_035119               1819 bp   ...   \n",
       "9457     complete cds  LOCUS       NC_034631               4461 bp   ...   \n",
       "9458     complete cds  LOCUS       NC_034633               6873 bp   ...   \n",
       "9459     complete cds  LOCUS       NC_034632                978 bp   ...   \n",
       "\n",
       "                                               Sequence  \n",
       "9455  tatactacaatatatacaatatttttcactcataaaatgactgact...  \n",
       "9456  ggcaatatgaattcctatgcacttaacaacattattcgtacttcac...  \n",
       "9457  agtagtgtactaccaatattctaatattttcaatatatctttacaa...  \n",
       "9458  agtagtgtactcctatactaacttcacttatcactaaaatggacga...  \n",
       "9459  agtagtgtactccagaaaaaagacaataacaaatctcaatctacaa...  "
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "viral_genomes.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Extract the molecule type (RNA or DNA) from the report\n",
    "viral_genomes['Molecule Type'] = 'None'\n",
    "\n",
    "def extract_genome_type(report):\n",
    "    try:\n",
    "        report_mol_type = re.findall('mol_type=[A-Z,a-z,=,\" ]+', report)\n",
    "        report_mol_type = report_mol_type[0]\n",
    "        molecule_type = re.findall('\"[A-z ]+\"', report_mol_type)\n",
    "        molecule_type = molecule_type[0].strip('\"')\n",
    "        return(molecule_type)\n",
    "    except IndexError:\n",
    "        print('error at', report)\n",
    "        pass\n",
    "    \n",
    "viral_genomes['Molecule Type'] = viral_genomes['Report'].apply(extract_genome_type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Extract the phylogeny from the report\n",
    "viral_genomes['Phylogeny'] = 'None'\n",
    "\n",
    "def extract_phylogeny(report):\n",
    "    try:\n",
    "        report_phylo = re.findall('Viruses[A-z ;,]+', report)\n",
    "        report_phylo = report_phylo[0]\n",
    "        return(report_phylo)\n",
    "    except IndexError:\n",
    "        print('error at', report)\n",
    "        pass\n",
    "    \n",
    "viral_genomes['Phylogeny'] = viral_genomes['Report'].apply(extract_phylogeny)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 9460 entries, 0 to 9459\n",
      "Data columns (total 19 columns):\n",
      "Genome                9460 non-null object\n",
      "Accession             9460 non-null object\n",
      "Source information    9460 non-null object\n",
      "Segm                  9460 non-null object\n",
      "Length                9460 non-null object\n",
      "Protein               9460 non-null object\n",
      "Neighbors             9460 non-null object\n",
      "Host                  9460 non-null object\n",
      "Created               9460 non-null object\n",
      "Updated               9460 non-null object\n",
      "Classification        9460 non-null object\n",
      "Extra Info            3000 non-null object\n",
      "BeautifulSoup         9460 non-null object\n",
      "ItemId                9460 non-null object\n",
      "SequenceType          9460 non-null object\n",
      "Report                9460 non-null object\n",
      "Sequence              9460 non-null object\n",
      "Molecule Type         9460 non-null object\n",
      "Phylogeny             9460 non-null object\n",
      "dtypes: object(19)\n",
      "memory usage: 1.4+ MB\n"
     ]
    }
   ],
   "source": [
    "viral_genomes.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "viral_genomes.to_csv('full_viral_genomes.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Source information</th>\n",
       "      <th>Segm</th>\n",
       "      <th>Length</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Neighbors</th>\n",
       "      <th>Host</th>\n",
       "      <th>Created</th>\n",
       "      <th>Updated</th>\n",
       "      <th>Classification</th>\n",
       "      <th>Extra Info</th>\n",
       "      <th>BeautifulSoup</th>\n",
       "      <th>ItemId</th>\n",
       "      <th>SequenceType</th>\n",
       "      <th>Report</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Molecule Type</th>\n",
       "      <th>Phylogeny</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9455</th>\n",
       "      <td>Wilkie partiti-like virus 1</td>\n",
       "      <td>/nuccore/1211677385</td>\n",
       "      <td>strain:mosWSCP36002</td>\n",
       "      <td>-</td>\n",
       "      <td>2100nt</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/14/2017</td>\n",
       "      <td>unclassified viruses</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\\n&lt;!DOCT...</td>\n",
       "      <td>NC_035122.1</td>\n",
       "      <td>complete genome</td>\n",
       "      <td>LOCUS       NC_035122               2100 bp   ...</td>\n",
       "      <td>tatactacaatatatacaatatttttcactcataaaatgactgact...</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>Viruses; unclassified viruses</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9456</th>\n",
       "      <td>Wilkie partiti-like virus 2</td>\n",
       "      <td>/nuccore/1211677379</td>\n",
       "      <td>strain:mosWSCP53020</td>\n",
       "      <td>-</td>\n",
       "      <td>1819nt</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>eukaryotes</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/14/2017</td>\n",
       "      <td>unclassified viruses</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\\n&lt;!DOCT...</td>\n",
       "      <td>NC_035119.1</td>\n",
       "      <td>complete genome</td>\n",
       "      <td>LOCUS       NC_035119               1819 bp   ...</td>\n",
       "      <td>ggcaatatgaattcctatgcacttaacaacattattcgtacttcac...</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>Viruses; unclassified viruses</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9457</th>\n",
       "      <td>Wolkberg virus segment 1</td>\n",
       "      <td>/nuccore/1197509331</td>\n",
       "      <td>isolate:2562_SA3</td>\n",
       "      <td>None</td>\n",
       "      <td>(4461 nt)</td>\n",
       "      <td>proteins: 1</td>\n",
       "      <td>neighbors: 10</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>05/24/2017</td>\n",
       "      <td>05/24/2017</td>\n",
       "      <td>unclassified viruses</td>\n",
       "      <td>['Wolkberg virus', '', 'isolate:2562_SA3', '3'...</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\\n&lt;!DOCT...</td>\n",
       "      <td>NC_034631.1</td>\n",
       "      <td>complete cds</td>\n",
       "      <td>LOCUS       NC_034631               4461 bp   ...</td>\n",
       "      <td>agtagtgtactaccaatattctaatattttcaatatatctttacaa...</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>Viruses; unclassified viruses</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9458</th>\n",
       "      <td>Wolkberg virus segment 2</td>\n",
       "      <td>/nuccore/1197509335</td>\n",
       "      <td>isolate:2562_SA3</td>\n",
       "      <td>None</td>\n",
       "      <td>(6873 nt)</td>\n",
       "      <td>proteins: 1</td>\n",
       "      <td>neighbors: 10</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>05/24/2017</td>\n",
       "      <td>05/24/2017</td>\n",
       "      <td>unclassified viruses</td>\n",
       "      <td>['Wolkberg virus', '', 'isolate:2562_SA3', '3'...</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\\n&lt;!DOCT...</td>\n",
       "      <td>NC_034633.1</td>\n",
       "      <td>complete cds</td>\n",
       "      <td>LOCUS       NC_034633               6873 bp   ...</td>\n",
       "      <td>agtagtgtactcctatactaacttcacttatcactaaaatggacga...</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>Viruses; unclassified viruses</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9459</th>\n",
       "      <td>Wolkberg virus segment 3</td>\n",
       "      <td>/nuccore/1197509333</td>\n",
       "      <td>isolate:2562_SA3</td>\n",
       "      <td>None</td>\n",
       "      <td>(978 nt)</td>\n",
       "      <td>proteins: 1</td>\n",
       "      <td>neighbors: 10</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>05/24/2017</td>\n",
       "      <td>05/24/2017</td>\n",
       "      <td>unclassified viruses</td>\n",
       "      <td>['Wolkberg virus', '', 'isolate:2562_SA3', '3'...</td>\n",
       "      <td>&lt;?xml version=\"1.0\" encoding=\"utf-8\"?&gt;\\n&lt;!DOCT...</td>\n",
       "      <td>NC_034632.1</td>\n",
       "      <td>complete cds</td>\n",
       "      <td>LOCUS       NC_034632                978 bp   ...</td>\n",
       "      <td>agtagtgtactccagaaaaaagacaataacaaatctcaatctacaa...</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>Viruses; unclassified viruses</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Genome            Accession   Source information  \\\n",
       "9455  Wilkie partiti-like virus 1  /nuccore/1211677385  strain:mosWSCP36002   \n",
       "9456  Wilkie partiti-like virus 2  /nuccore/1211677379  strain:mosWSCP53020   \n",
       "9457     Wolkberg virus segment 1  /nuccore/1197509331     isolate:2562_SA3   \n",
       "9458     Wolkberg virus segment 2  /nuccore/1197509335     isolate:2562_SA3   \n",
       "9459     Wolkberg virus segment 3  /nuccore/1197509333     isolate:2562_SA3   \n",
       "\n",
       "      Segm     Length      Protein      Neighbors           Host     Created  \\\n",
       "9455     -     2100nt            1              -  invertebrates  06/27/2017   \n",
       "9456     -     1819nt            1              1     eukaryotes  06/27/2017   \n",
       "9457  None  (4461 nt)  proteins: 1  neighbors: 10  invertebrates  05/24/2017   \n",
       "9458  None  (6873 nt)  proteins: 1  neighbors: 10  invertebrates  05/24/2017   \n",
       "9459  None   (978 nt)  proteins: 1  neighbors: 10  invertebrates  05/24/2017   \n",
       "\n",
       "         Updated        Classification  \\\n",
       "9455  07/14/2017  unclassified viruses   \n",
       "9456  07/14/2017  unclassified viruses   \n",
       "9457  05/24/2017  unclassified viruses   \n",
       "9458  05/24/2017  unclassified viruses   \n",
       "9459  05/24/2017  unclassified viruses   \n",
       "\n",
       "                                             Extra Info  \\\n",
       "9455                                                NaN   \n",
       "9456                                                NaN   \n",
       "9457  ['Wolkberg virus', '', 'isolate:2562_SA3', '3'...   \n",
       "9458  ['Wolkberg virus', '', 'isolate:2562_SA3', '3'...   \n",
       "9459  ['Wolkberg virus', '', 'isolate:2562_SA3', '3'...   \n",
       "\n",
       "                                          BeautifulSoup       ItemId  \\\n",
       "9455  <?xml version=\"1.0\" encoding=\"utf-8\"?>\\n<!DOCT...  NC_035122.1   \n",
       "9456  <?xml version=\"1.0\" encoding=\"utf-8\"?>\\n<!DOCT...  NC_035119.1   \n",
       "9457  <?xml version=\"1.0\" encoding=\"utf-8\"?>\\n<!DOCT...  NC_034631.1   \n",
       "9458  <?xml version=\"1.0\" encoding=\"utf-8\"?>\\n<!DOCT...  NC_034633.1   \n",
       "9459  <?xml version=\"1.0\" encoding=\"utf-8\"?>\\n<!DOCT...  NC_034632.1   \n",
       "\n",
       "         SequenceType                                             Report  \\\n",
       "9455  complete genome  LOCUS       NC_035122               2100 bp   ...   \n",
       "9456  complete genome  LOCUS       NC_035119               1819 bp   ...   \n",
       "9457     complete cds  LOCUS       NC_034631               4461 bp   ...   \n",
       "9458     complete cds  LOCUS       NC_034633               6873 bp   ...   \n",
       "9459     complete cds  LOCUS       NC_034632                978 bp   ...   \n",
       "\n",
       "                                               Sequence Molecule Type  \\\n",
       "9455  tatactacaatatatacaatatttttcactcataaaatgactgact...   genomic RNA   \n",
       "9456  ggcaatatgaattcctatgcacttaacaacattattcgtacttcac...   genomic RNA   \n",
       "9457  agtagtgtactaccaatattctaatattttcaatatatctttacaa...   genomic RNA   \n",
       "9458  agtagtgtactcctatactaacttcacttatcactaaaatggacga...   genomic RNA   \n",
       "9459  agtagtgtactccagaaaaaagacaataacaaatctcaatctacaa...   genomic RNA   \n",
       "\n",
       "                          Phylogeny  \n",
       "9455  Viruses; unclassified viruses  \n",
       "9456  Viruses; unclassified viruses  \n",
       "9457  Viruses; unclassified viruses  \n",
       "9458  Viruses; unclassified viruses  \n",
       "9459  Viruses; unclassified viruses  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## read in the full dataframe.\n",
    "chunksize = 1000\n",
    "TextFileReader = pd.read_csv('full_viral_genomes.csv', chunksize=chunksize, iterator=True)\n",
    "viral_genomes = pd.concat(TextFileReader, ignore_index=True)\n",
    "\n",
    "viral_genomes.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "viral_genomes = viral_genomes[[col for col in viral_genomes.columns if col != 'BeautifulSoup' and col != 'Extra Info']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Source information</th>\n",
       "      <th>Segm</th>\n",
       "      <th>Length</th>\n",
       "      <th>Protein</th>\n",
       "      <th>Neighbors</th>\n",
       "      <th>Host</th>\n",
       "      <th>Created</th>\n",
       "      <th>Updated</th>\n",
       "      <th>Classification</th>\n",
       "      <th>ItemId</th>\n",
       "      <th>SequenceType</th>\n",
       "      <th>Report</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Molecule Type</th>\n",
       "      <th>Phylogeny</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hepatitis delta virus</td>\n",
       "      <td>/nuccore/13277517</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>1682nt</td>\n",
       "      <td>2</td>\n",
       "      <td>320</td>\n",
       "      <td>vertebrates, human</td>\n",
       "      <td>04/29/1993</td>\n",
       "      <td>02/10/2015</td>\n",
       "      <td>Deltavirus</td>\n",
       "      <td>NC_001653.2</td>\n",
       "      <td>complete genome</td>\n",
       "      <td>LOCUS       NC_001653               1682 bp ss...</td>\n",
       "      <td>atgagccaagttccgaacaaggattcgcggggaggatagatcagcg...</td>\n",
       "      <td>genomic RNA</td>\n",
       "      <td>Viruses; Deltavirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Badger associated gemykibivirus 1</td>\n",
       "      <td>/nuccore/807743872</td>\n",
       "      <td>strain:588t</td>\n",
       "      <td>-</td>\n",
       "      <td>2112nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>vertebrates</td>\n",
       "      <td>04/16/2015</td>\n",
       "      <td>11/24/2015</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>NC_026806.1</td>\n",
       "      <td>complete genome</td>\n",
       "      <td>LOCUS       NC_026806               2112 bp   ...</td>\n",
       "      <td>taatactatagccatctgggacacagagcacagtcgacgtgtccct...</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>Viruses; Genomoviridae; Gemykibivirus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bemisia-associated genomovirus AdDF</td>\n",
       "      <td>/nuccore/1211677462</td>\n",
       "      <td>isolate:AdDF</td>\n",
       "      <td>-</td>\n",
       "      <td>2199nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/19/2017</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>NC_035137.1</td>\n",
       "      <td>complete genome</td>\n",
       "      <td>LOCUS       NC_035137               2199 bp   ...</td>\n",
       "      <td>taatgttatacacaaaccgtggtgtcagtgtcacctctatatagta...</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>Viruses; Genomoviridae; unclassified Genomovir...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bemisia-associated genomovirus AdO</td>\n",
       "      <td>/nuccore/1211677465</td>\n",
       "      <td>isolate:AdO</td>\n",
       "      <td>-</td>\n",
       "      <td>2211nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/19/2017</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>NC_035138.1</td>\n",
       "      <td>complete genome</td>\n",
       "      <td>LOCUS       NC_035138               2211 bp   ...</td>\n",
       "      <td>taatattatagcccaggacacagggcacacctgtgtctatataaac...</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>Viruses; Genomoviridae; unclassified Genomovir...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bemisia-associated genomovirus NfO</td>\n",
       "      <td>/nuccore/1211677468</td>\n",
       "      <td>isolate:NfO</td>\n",
       "      <td>-</td>\n",
       "      <td>2231nt</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>invertebrates</td>\n",
       "      <td>06/27/2017</td>\n",
       "      <td>07/19/2017</td>\n",
       "      <td>Genomoviridae</td>\n",
       "      <td>NC_035139.1</td>\n",
       "      <td>complete genome</td>\n",
       "      <td>LOCUS       NC_035139               2231 bp   ...</td>\n",
       "      <td>taatattattctctctctcaggcagaggtagagggagttaatatag...</td>\n",
       "      <td>genomic DNA</td>\n",
       "      <td>Viruses; Genomoviridae; unclassified Genomovir...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Genome            Accession  \\\n",
       "0                Hepatitis delta virus    /nuccore/13277517   \n",
       "1    Badger associated gemykibivirus 1   /nuccore/807743872   \n",
       "2  Bemisia-associated genomovirus AdDF  /nuccore/1211677462   \n",
       "3   Bemisia-associated genomovirus AdO  /nuccore/1211677465   \n",
       "4   Bemisia-associated genomovirus NfO  /nuccore/1211677468   \n",
       "\n",
       "  Source information Segm  Length Protein Neighbors                Host  \\\n",
       "0                NaN    -  1682nt       2       320  vertebrates, human   \n",
       "1        strain:588t    -  2112nt       2         -         vertebrates   \n",
       "2       isolate:AdDF    -  2199nt       2         -       invertebrates   \n",
       "3        isolate:AdO    -  2211nt       2         -       invertebrates   \n",
       "4        isolate:NfO    -  2231nt       2         -       invertebrates   \n",
       "\n",
       "      Created     Updated Classification       ItemId     SequenceType  \\\n",
       "0  04/29/1993  02/10/2015     Deltavirus  NC_001653.2  complete genome   \n",
       "1  04/16/2015  11/24/2015  Genomoviridae  NC_026806.1  complete genome   \n",
       "2  06/27/2017  07/19/2017  Genomoviridae  NC_035137.1  complete genome   \n",
       "3  06/27/2017  07/19/2017  Genomoviridae  NC_035138.1  complete genome   \n",
       "4  06/27/2017  07/19/2017  Genomoviridae  NC_035139.1  complete genome   \n",
       "\n",
       "                                              Report  \\\n",
       "0  LOCUS       NC_001653               1682 bp ss...   \n",
       "1  LOCUS       NC_026806               2112 bp   ...   \n",
       "2  LOCUS       NC_035137               2199 bp   ...   \n",
       "3  LOCUS       NC_035138               2211 bp   ...   \n",
       "4  LOCUS       NC_035139               2231 bp   ...   \n",
       "\n",
       "                                            Sequence Molecule Type  \\\n",
       "0  atgagccaagttccgaacaaggattcgcggggaggatagatcagcg...   genomic RNA   \n",
       "1  taatactatagccatctgggacacagagcacagtcgacgtgtccct...   genomic DNA   \n",
       "2  taatgttatacacaaaccgtggtgtcagtgtcacctctatatagta...   genomic DNA   \n",
       "3  taatattatagcccaggacacagggcacacctgtgtctatataaac...   genomic DNA   \n",
       "4  taatattattctctctctcaggcagaggtagagggagttaatatag...   genomic DNA   \n",
       "\n",
       "                                           Phylogeny  \n",
       "0                                Viruses; Deltavirus  \n",
       "1              Viruses; Genomoviridae; Gemykibivirus  \n",
       "2  Viruses; Genomoviridae; unclassified Genomovir...  \n",
       "3  Viruses; Genomoviridae; unclassified Genomovir...  \n",
       "4  Viruses; Genomoviridae; unclassified Genomovir...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "viral_genomes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "viral_genomes.to_csv('final_viral_genomes.csv', index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
